{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### SETTING UP THE DATAFRAME ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1151,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Importing relevant modules: \n",
    "# * dataframe - pandas \n",
    "# * graphs - matplotlib \n",
    "# * tools - numpy, random, fnmatch, os\n",
    "\n",
    "\n",
    "# * fasta to dataframe conversion - SeqIO\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import os\n",
    "import pandas as pd\n",
    "import random\n",
    "import collections\n",
    "import math\n",
    "from pandas.tools.plotting import table\n",
    "from itertools import product\n",
    "from Bio import SeqIO\n",
    "from ete3 import NCBITaxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1535,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#define root directory within which to create a subdirectory system\n",
    "#define input paths for basecalled read fasta files, rgbesthit and ncbibesthit files\n",
    "work_dir = '/home/gamran/dirtreetest/'\n",
    "fasta_input_dir_path = '/'\n",
    "rgbesthit_input_dir_path = '/' \n",
    "ncbibesthit_input_dir_path = '/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1645,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "good_barcodes = ['BC01', 'BC02', 'BC03', 'BC04', 'BC05', 'BC06']\n",
    "ref_genome_tag_list = ['TGAC', 'Zymo', 'Ptri', 'Pst', 'tig']\n",
    "ref_genome_species_list = ['Wheat', \n",
    "                        'Zymoseptoria tritici', \n",
    "                        'Pyrenophore tritici-repentis', \n",
    "                        'P. striformis f. sp. tritici WA', \n",
    "                        'Parastagonospora nodorum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Functions that generate main work directories and subdirectories for:\n",
    "# * data (all fasta files)\n",
    "# * blast (all blast output files)\n",
    "# * analysis (all dataframe summary and graph files)\n",
    "# When given a top directory path\n",
    "\n",
    "def create_main_directories(top_dir):\n",
    "    \"\"\"Given a top work directory, creates 3 main subdirectories - data, blast and analysis\"\"\"\n",
    "    main_directory_list = ['data/', 'blast/', 'analysis/']\n",
    "    for new_direct in main_directory_list:\n",
    "        if not os.path.exists(top_dir + new_direct):\n",
    "            os.makedirs(top_dir + new_direct)\n",
    "\n",
    "def create_data_subdirectories(top_dir):\n",
    "    data_subdirectory_list = ['basecalled/',\n",
    "                        'rghityes/',\n",
    "                        'rghitno/',\n",
    "                        'rghitsamples/']\n",
    "    data_subsubdirectory_list = ['ncbihityes/',\n",
    "                                 'ncbihitno/']\n",
    "    for new_subdirect in data_subdirectory_list:\n",
    "        if not os.path.exists(top_dir + 'data/' + new_subdirect):\n",
    "            os.makedirs(top_dir + 'data/' + new_subdirect)\n",
    "    for new_subsubdirect in data_subsubdirectory_list:\n",
    "        if not os.path.exists(top_dir + 'data/rghitno/' + new_subsubdirect):\n",
    "            os.makedirs(top_dir + 'data/rghitno/' + new_subsubdirect)\n",
    "\n",
    "def create_blast_subdirectories(top_dir):\n",
    "    blast_subdirectory_list = ['rgbesthit/',\n",
    "                               'ncbibesthit/']\n",
    "    for new_subdirect in blast_subdirectory_list:\n",
    "        if not os.path.exists(top_dir + 'blast/' + new_subdirect):\n",
    "            os.makedirs(top_dir + 'blast/' + new_subdirect)\n",
    "\n",
    "def create_analysis_subdirectories(top_dir):\n",
    "    analysis_subdirectory_list = ['graphs/',\n",
    "                                  'tables/',\n",
    "                               'summaries/']\n",
    "    for new_subdirect in analysis_subdirectory_list:\n",
    "        if not os.path.exists(top_dir + 'analysis/' + new_subdirect):\n",
    "            os.makedirs(top_dir + 'analysis/' + new_subdirect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Apply previous functions to generate data, blast and analysis directories and subdirectories:\n",
    "create_main_directories(work_dir)\n",
    "create_data_subdirectories(work_dir)\n",
    "create_blast_subdirectories(work_dir)\n",
    "create_analysis_subdirectories(work_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1572,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Reduces basecalled_file_names to a flattened list, extracts all basecalled fasta file names\n",
    "def flatten(l):\n",
    "    for el in l:\n",
    "        if isinstance(el, collections.Iterable) and not isinstance(el, (str, bytes)):\n",
    "            yield from flatten(el)\n",
    "        else:\n",
    "            yield el\n",
    "            \n",
    "def get_file_type(string, delimiter):\n",
    "    file_type = string.split(delimiter)\n",
    "    return delimiter + file_type[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1602,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Extracts fasta files from the initial input folder and moves them to data/basecalled to be manipulated\n",
    "fasta_input_file_names = os.listdir(fasta_input_dir_path)\n",
    "fasta_input_file_names.sort()\n",
    "if fasta_input_file_names:\n",
    "    fasta_file_type = get_file_type(fasta_input_file_names[-1], '.')\n",
    "\n",
    "    for fasta_file in fasta_input_file_names:\n",
    "        if os.path.exists(fasta_input_dir_path + fasta_file):\n",
    "            if not os.path.exists(work_dir + 'data/basecalled/' + fasta_file):\n",
    "                os.rename(fasta_input_dir_path + fasta_file, work_dir + 'data/basecalled/' + fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1601,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Generates a list of all the fasta files present in data/basecalled\n",
    "basecalled_file_names = os.listdir(work_dir + 'data/basecalled')\n",
    "basecalled_fasta_file_names = [x for x in basecalled_file_names if fasta_file_type in x]\n",
    "basecalled_fasta_file_names.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 835,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Make a list of all available barcodes the basecalled reads are sorted into, based on fasta file names\n",
    "basecalled_fasta_barcodes = []\n",
    "for bcs in range(0, len(basecalled_fasta_file_names)):\n",
    "    bclabel_index = ''\n",
    "    if int(basecalled_fasta_file_names[bcs].rfind('BC')) > int(basecalled_fasta_file_names[bcs].rfind('NB')):\n",
    "        bclabel_index = int(basecalled_fasta_file_names[bcs].rfind('BC'))\n",
    "    else:\n",
    "        bclabel_index = int(basecalled_fasta_file_names[bcs].rfind('NB'))\n",
    "    basecalled_fasta_barcodes.append(basecalled_fasta_file_names[bcs][bclabel_index:bclabel_index + 4]) #Convert to a search parameter? Try this Yiheng!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Specifies all barcodes and bad barcodes (good_barcodes is previously defined)\n",
    "all_barcodes = basecalled_fasta_barcodes\n",
    "bad_barcodes = sorted(list(set(all_barcodes) - set(good_barcodes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 928,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Extracts the ID from the basecalled file names (everything before BC/NB??.fa)\n",
    "basecalled_fasta_file_id_list = []\n",
    "for file_name in basecalled_fasta_file_names:\n",
    "    bclabel_index = ''\n",
    "    if int(file_name.rfind('BC')) > int(file_name.rfind('NB')):\n",
    "        bclabel_index = int(file_name.rfind('BC'))\n",
    "        basecalled_fasta_file_id_list.append(file_name[0:bclabel_index])\n",
    "    else:\n",
    "        bclabel_index = int(file_name.rfind('NB'))\n",
    "        basecalled_fasta_file_id_list.append(file_name[0:bclabel_index])\n",
    "    \n",
    "basecalled_fasta_file_id = list(set(basecalled_fasta_file_id_list))[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Create text file to summarise read_id, barcode and length for all basecalled reads, save in analysis/summaries/\n",
    "analysis_summaries_path = work_dir + 'analysis/summaries/'\n",
    "complete_path_and_name = os.path.join(analysis_summaries_path, basecalled_fasta_file_id + \".lengths.txt\")\n",
    "lengths_txt = open(complete_path_and_name, \"w\")\n",
    "print(\"Read_id,Barcode,Length,Quality\", file=lengths_txt)\n",
    "\n",
    "os.chdir(work_dir + 'data/basecalled/')\n",
    "for bcs, fa_file in zip(basecalled_fasta_barcodes, basecalled_fasta_file_names):\n",
    "    for seq in SeqIO.parse(open(fa_file), 'fasta'):\n",
    "        if bcs in good_barcodes:\n",
    "            print(seq.id + \",\" + bcs + \",\" + str(len(seq)) + \",\" + 'Good', file=lengths_txt)\n",
    "        elif bcs in bad_barcodes:\n",
    "            print(seq.id + \",\" + 'NB00' + \",\" + str(len(seq)) + \",\" + 'Bad', file=lengths_txt)\n",
    "        else:\n",
    "            continue\n",
    "lengths_txt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "basecalled_df = pd.read_csv(complete_path_and_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### DATAFRAME ANALYSIS - TABLE OF SUMMARY DATA ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "### Shows number of reads, sum, max, mean, median of read lengths per barcode, all bad barcodes sorted to NB00 ###\n",
    "basecalled_df_pivot = basecalled_df.pivot_table(values='Length', \n",
    "                                                index='Barcode', \n",
    "                                                aggfunc=[len, np.sum, np.max, np.mean, np.median],\n",
    "                                                fill_value=0, \n",
    "                                                margins=True)\n",
    "basecalled_df_pivot.index.name = None\n",
    "\n",
    "for clmns in ['len', 'amax', 'median']:\n",
    "    basecalled_df_pivot[clmns] = basecalled_df_pivot[clmns].apply(lambda x: int(x))\n",
    "basecalled_df_pivot['sum'] = basecalled_df_pivot['sum'].apply(lambda x: \"{0:.3f}\".format(int(x)/1000000))\n",
    "basecalled_df_pivot['mean'] = basecalled_df_pivot['mean'].apply(lambda x: \"{0:.2f}\".format(x))\n",
    "basecalled_df_pivot_formal = basecalled_df_pivot.copy() #titles have whitespace, make fancier\n",
    "basecalled_df_pivot_formal.rename(columns={'len': 'Number of Reads', \n",
    "                                    'sum': 'Total Length (Mbp)',\n",
    "                                    'amax': 'Max Length (bp)',\n",
    "                                    'mean': 'Mean Length (bp)',\n",
    "                                    'median': 'Median Length (bp)'}, inplace=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 437,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "### Saves pivot table summary in analysis/tables/ folder ###\n",
    "fig = plt.figure(figsize=(9,2))\n",
    "ax = fig.add_subplot(1, 1, 1, frame_on=False)\n",
    "ax.xaxis.set_visible(False)  # hide the x axis\n",
    "ax.yaxis.set_visible(False)  # hide the y axis\n",
    "\n",
    "table(ax, basecalled_df_pivot_formal, loc='center')\n",
    "\n",
    "plt.savefig(os.path.join(work_dir + 'analysis/tables/', \n",
    "                        basecalled_fasta_file_id + \"_basecalled_reads_summary.png\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "### DATAFRAME ANALYSIS - HISTOGRAMS OF SUMMARY DATA ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def random_color(x):\n",
    "    r = lambda: random.randint(0,255)\n",
    "    return ('#%02X%02X%02X' % (r(),r(),r()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1517,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Generates a histogram showing the total read length distribution for all basecalled reads\n",
    "plt.figure(figsize=(10,5))\n",
    "basecalled_df.Length.hist(bins=30)\n",
    "\n",
    "plt.title('Basecalled Reads - Total Read Length Distribution', y=1.03, fontsize='x-large', fontweight='bold')\n",
    "\n",
    "plt.xlabel('Read Length', fontsize=16)\n",
    "plt.xticks(np.arange(0, basecalled_df.Length.max() + 1, 2000))\n",
    "plt.xlim([-1000, basecalled_df.Length.max() + 1000])\n",
    "\n",
    "plt.ylabel('Read Count', fontsize=16)\n",
    "\n",
    "for idx, clmn_name in enumerate(list(basecalled_df_pivot_formal.columns)):\n",
    "    plt.annotate(clmn_name + ' = ' + str(basecalled_df_pivot_formal[clmn_name]['All']), \n",
    "                 xy=(1, 1), \n",
    "                 xycoords='axes fraction', \n",
    "                 fontsize=16, \n",
    "                 fontweight='normal',\n",
    "                 xytext=(-20, -30 - 30*idx), \n",
    "                 textcoords='offset points', \n",
    "                 ha='right', \n",
    "                 va='top')\n",
    "    \n",
    "plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'basecalled_read_distribution_total.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1668,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "#Generates a histogram showing the total read length distribution for all basecalled reads per barcode\n",
    "\n",
    "#Assesses number of barcodes, generates necessary number of plots (even) and relevant indexing system\n",
    "basecalled_barcodes = list(basecalled_df['Barcode'].unique())\n",
    "\n",
    "if len(list(basecalled_df['Barcode'].unique())) % 2 == 0:\n",
    "    no_of_subplots = len(basecalled_barcodes)\n",
    "else:\n",
    "    no_of_subplots = len(basecalled_barcodes) + 1\n",
    "\n",
    "#Always 2 columns, bc-count/2 rows \n",
    "no_of_subplots_pair = [int(no_of_subplots/2), 2]\n",
    "\n",
    "#Produce pairs of indices correlating to the coordinates of the subplots\n",
    "subplot_coordinates = list(product(range(no_of_subplots_pair[0]), range(no_of_subplots_pair[1])))\n",
    "subplot_coordinates_list = [list(l) for l in subplot_coordinates]\n",
    "subplot_coordinates_list_rows = [i[0] for i in subplot_coordinates_list]\n",
    "subplot_coordinates_list_columns = [i[1] for i in subplot_coordinates_list]\n",
    "\n",
    "fig, ax = plt.subplots(no_of_subplots_pair[0], no_of_subplots_pair[1], figsize=(15,10))\n",
    "\n",
    "xmax = int(basecalled_df.Length.max())\n",
    "\n",
    "max_count_list = []\n",
    "for bcs in list(basecalled_df['Barcode'].unique()):\n",
    "    max_count_list.append(np.histogram(basecalled_df.loc[basecalled_df['Barcode'] == bcs]['Length'], 60)[0].max())\n",
    "max_count_list.sort()\n",
    "max_count_list_max = max_count_list[-1]\n",
    "max_count_list_max_digits = len(str(max_count_list_max))\n",
    "exact_ylim = (int(max_count_list_max / 10**(max_count_list_max_digits-1))+1)*(10**(max_count_list_max_digits-1))\n",
    "\n",
    "def applyGroupHistograms(ax_ind1, ax_ind2, bcs):\n",
    "    ax[ax_ind1, ax_ind2].hist(basecalled_df.groupby('Barcode')['Length'].get_group(bcs), \n",
    "                              bins=range(0, xmax, int(xmax/60)), \n",
    "                              color=random_color(x), \n",
    "                              alpha=0.8)\n",
    "    ax[ax_ind1, ax_ind2].set_title(bcs)\n",
    "    ax[ax_ind1, ax_ind2].set_xlabel('Read Length')\n",
    "    ax[ax_ind1, ax_ind2].set_ylabel('Read Count');\n",
    "    ax[ax_ind1, ax_ind2].set_xlim([-1000, basecalled_df['Length'].max() + 1000])\n",
    "    ax[ax_ind1, ax_ind2].set_xticks(np.arange(0, xmax + 1, 2000))\n",
    "    ax[ax_ind1, ax_ind2].set_ylim(0, exact_ylim)\n",
    "    ax[ax_ind1, ax_ind2].grid(True, which='Major')\n",
    "    \n",
    "    for idx, clmn_name in enumerate(['Number of Reads', 'Total Length (Mbp)', 'Median Length (bp)']):\n",
    "        ax[ax_ind1, ax_ind2].annotate(clmn_name + ' = ' + str(basecalled_df_pivot_formal[clmn_name][bcs]), \n",
    "                     xy=(1, 1), \n",
    "                     xycoords='axes fraction', \n",
    "                     fontsize=16, \n",
    "                     fontweight='normal',\n",
    "                     xytext=(-20, -25 - 30*idx), \n",
    "                     textcoords='offset points', \n",
    "                     ha='right', \n",
    "                     va='top')\n",
    "\n",
    "for ax_ind1, ax_ind2, bcs, in zip(subplot_coordinates_list_rows, subplot_coordinates_list_columns, bc_list):\n",
    "    applyGroupHistograms(ax_ind1, ax_ind2, bcs)    \n",
    "\n",
    "if len(list(basecalled_df['Barcode'].unique())) != 0:\n",
    "    plt.delaxes(ax[subplot_coordinates_list_rows[-1], subplot_coordinates_list_columns[-1]])\n",
    "\n",
    "plt.suptitle('Basecalled Reads - Read Length Distribution By Barcode', \n",
    "             y=1.03, \n",
    "             fontsize='x-large', \n",
    "             fontweight='bold')\n",
    "plt.tight_layout()\n",
    "plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'basecalled_read_distribution_bcsort.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### DATAFRAME ANALYSIS - COLLATION OF RGBESTHIT DATA ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1604,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Extracts rgblast files from the initial input folder and moves them to data/basecalled to be manipulated\n",
    "rgblast_input_file_names = os.listdir(rgblast_input_dir_path)\n",
    "rgblast_input_file_names.sort()\n",
    "if rgblast_input_file_names:\n",
    "    rgblast_file_type = get_file_type(rgblast_input_file_names[-1], '.')\n",
    "\n",
    "    for rgblast_file in rgblast_input_file_names:\n",
    "        if os.path.exists(rgblast_input_dir_path + rgblast_file):\n",
    "            if not os.path.exists(work_dir + 'blast/rgbesthit/' + rgblast_file):\n",
    "                os.rename(rgblast_input_dir_path + rgblast_file, work_dir + 'blast/rgbesthit/' + rgblast_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1605,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# creates a list of all the reference genome blast besthit file names in blast/rgbesthit\n",
    "rgblast_file_names = os.listdir(work_dir + 'blast/rgbesthit')\n",
    "rgblast_besthit_file_names = [x for x in rgblast_file_names if rgblast_file_type in x]\n",
    "rgblast_besthit_file_names.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1606,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Extracts the reference genome blast id from the rgblast hit file names\n",
    "rgblast_file_id_list = []\n",
    "for file_name in rgblast_besthit_file_names:\n",
    "    start = file_name.find(fasta_file_type)\n",
    "    end = file_name.find(rgblast_file_type)\n",
    "    rgblast_file_id_list.append(file_name[start+len(fasta_file_type):end])\n",
    "rgblast_file_id = list(set(rgblast_file_id_list))[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1625,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Make a list of all available barcodes the rgbesthit reads are sorted into, based on rgbesthit file names\n",
    "blast_besthit_file_names_idstrip = [file_name.strip(basecalled_fasta_file_id) for file_name in rgblast_besthit_file_names]\n",
    "rgblast_besthit_barcodes = []\n",
    "for bcs in range(0, len(rgblast_besthit_file_names_idstrip)):\n",
    "    bclabel_index = ''\n",
    "    if int(rgblast_besthit_file_names_idstrip[bcs].find('BC')) > int(rgblast_besthit_file_names_idstrip[bcs].find('NB')):\n",
    "        bclabel_index = int(rgblast_besthit_file_names_idstrip[bcs].find('BC'))\n",
    "    else:\n",
    "        bclabel_index = int(rgblast_besthit_file_names_idstrip[bcs].find('NB'))\n",
    "    rgblast_besthit_barcodes.append(rgblast_besthit_file_names_idstrip[bcs][bclabel_index:bclabel_index + 4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1643,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Makes a dictionary - key (barcode) : value (dataframe of rgblast hit results per barcode)\n",
    "os.chdir(work_dir + 'blast/rgbesthit')\n",
    "rgblast_dfs_dict = {}\n",
    "blast_output_parameters = ['qseqid', 'sseqid', 'evalue', 'bitscore', 'length', 'pident', 'nident']\n",
    "for bcs, besthit_file in zip(rgblast_besthit_barcodes, rgblast_besthit_file_names):\n",
    "    if rgblast_file_id in besthit_file:\n",
    "        rgblast_dfs_dict[bcs] = pd.read_table(besthit_file, delim_whitespace=True, \n",
    "                                           names=blast_output_parameters) #hardcoded\n",
    "for bcs in rgblast_dfs_dict.keys():\n",
    "    rgblast_dfs_dict[bcs]['barcode'] = bcs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1644,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# NECESSARY TO HARD CODE - Must supply two lists of equal length, in the desired order: \n",
    "# one corresponding to unique letter combos in the sseqid ids (tags)\n",
    "# one corresponding to the full names of the reference genomes used\n",
    "\n",
    "def assign_species_from_sseqid(x):\n",
    "    for rg_tag, rg_full in zip(ref_genome_tag_list, ref_genome_species_list):    \n",
    "        if rg_tag in str(x):\n",
    "            return rg_full\n",
    "    else:\n",
    "        return 'N/A'\n",
    "\n",
    "for bcs in rgblast_dfs_dict.keys():\n",
    "    rgblast_dfs_dict[bcs]['species'] = rgblast_dfs_dict[bcs]['sseqid'].apply(assign_species_from_sseqid)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 566,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Combining the BC00 reads from BLAST\n",
    "if 'BC00' in rgblast_dfs_dict.keys():\n",
    "    rgblast_dfs_dict['BC00']['barcode'] = 'NB00'\n",
    "\n",
    "#Combine all the reference genome hit dataframes across all barcodes\n",
    "rgblast_df_list = []\n",
    "for bcs in rgblast_dfs_dict.keys():\n",
    "    rgblast_df_list.append(rgblast_dfs_dict[bcs])\n",
    "    \n",
    "rgblast_df_joint = pd.concat(rgblast_df_list, axis=0, ignore_index=True)\n",
    "rgblast_df_joint.sort_values('barcode', ascending=True, inplace=True)\n",
    "rgblast_df_joint.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 571,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Create a pivot table showing all reference genome hit counts for each barcode\n",
    "rgblast_df_joint_pivot = rgblast_df_joint.pivot_table(values='nident', index='barcode', columns='species', aggfunc='count', fill_value=0, margins=True)\n",
    "rgblast_df_joint_pivot = rgblast_df_joint_pivot.applymap(int)\n",
    "rgblast_df_joint_pivot.index.name = None\n",
    "rgblast_df_joint_pivot.columns.name = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 587,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Save pivot table to analysis/tables/\n",
    "fig = plt.figure(figsize=(11,2))\n",
    "ax = fig.add_subplot(1, 1, 1, frame_on=False)\n",
    "ax.xaxis.set_visible(False)  # hide the x axis\n",
    "ax.yaxis.set_visible(False)  # hide the y axis\n",
    "\n",
    "table(ax, rgblast_df_joint_pivot, loc='center')\n",
    "\n",
    "plt.savefig(os.path.join(work_dir + 'analysis/tables/', \n",
    "                        basecalled_fasta_file_id + \"_rgblasthit_reads_summary.png\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 573,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Removes the 'all' column (for graphing purposes)\n",
    "rgblast_df_joint_pivot_noall = rgblast_df_joint_pivot.drop('All', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1520,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Construct a grouped bar chart showing the reference genome hit counts per barcode (log y axis)\n",
    "def generateGBGforRGHC(dropmaxspecies=False):\n",
    "    n_groups = len(rgblast_df_joint_pivot_noall.index.values)\n",
    "\n",
    "    max_genome = str(rgblast_df_joint_pivot_noall.ix['All'].idxmax(axis=1))\n",
    "\n",
    "    spc_list = list(rgblast_df_joint_pivot_noall.columns)\n",
    "    \n",
    "    if dropmaxspecies == True:\n",
    "        spc_list.remove(max_genome)\n",
    "\n",
    "    n_species_dict = {}\n",
    "    for spc in spc_list:\n",
    "        n_species_dict[spc] = tuple(rgblast_df_joint_pivot_noall[spc])\n",
    "\n",
    "    n_nohit = tuple(basecalled_df_pivot['len'].values - rgblast_df_joint_pivot['All'].values)    \n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(7,3))\n",
    "\n",
    "    index = np.arange(n_groups)\n",
    "    bar_width = 0.13\n",
    "    opacity = 0.4\n",
    "\n",
    "    if dropmaxspecies == False:\n",
    "        plt.bar(index - bar_width*2, \n",
    "            n_nohit, \n",
    "            bar_width,\n",
    "            alpha = opacity,\n",
    "            log = True,\n",
    "            color = 'k',\n",
    "            label = 'No Hit')\n",
    "\n",
    "        for idx, spc in enumerate(spc_list):\n",
    "            plt.bar(index - bar_width + bar_width*idx, \n",
    "                n_species_dict[spc], \n",
    "                bar_width,\n",
    "                alpha = opacity,\n",
    "                log = True,\n",
    "                color = random_color(x),\n",
    "                label = spc)\n",
    "    else:\n",
    "        for idx, spc in enumerate(spc_list):\n",
    "            plt.bar(index - bar_width + bar_width*idx, \n",
    "                n_species_dict[spc], \n",
    "                bar_width,\n",
    "                alpha = opacity,\n",
    "                color = random_color(x),\n",
    "                label = spc)\n",
    "            \n",
    "    def generate_upper_ylim(graph_max):\n",
    "        \"\"\"Takes a value, counts the digits and outputs a value corresponding to 10*digits\"\"\"\n",
    "        digits = len(str(graph_max))\n",
    "        round_up = 10**(digits)\n",
    "        return round_up\n",
    "\n",
    "    plt.xlabel('Barcode')\n",
    "    plt.xticks(index + bar_width*1.2, tuple(rgblast_df_joint_pivot_noall.index.values))\n",
    "\n",
    "    plt.ylabel('Number of Reads')\n",
    "    if dropmaxspecies == False:\n",
    "        plt.ylim(0, generate_upper_ylim(rgblast_df_joint_pivot['All'].values.max()))\n",
    "    else:\n",
    "        rgblast_df_joint_pivot_nomaxspecies = rgblast_df_joint_pivot_noall.copy()\n",
    "        rgblast_df_joint_pivot_nomaxspecies.drop(max_genome, axis=1, inplace=True)\n",
    "        rgblast_df_joint_pivot_nomaxspecies['All'] = rgblast_df_joint_pivot_nomaxspecies.sum(axis=1)\n",
    "        plt.ylim(0, (rgblast_df_joint_pivot_nomaxspecies['All'].values.max()+1)/2)\n",
    "\n",
    "    plt.title('Number of Reads Hitting Specific Reference Genomes', y=1.08)\n",
    "    plt.legend(loc='center left', bbox_to_anchor=(1.01, 0.5), fontsize=10)\n",
    "    plt.grid(b=None, which='major', axis='y', linestyle='dashed', linewidth=0.2)\n",
    "    plt.margins(x=0.1)\n",
    "    plt.tick_params(top=\"off\", right=\"off\", bottom='off')\n",
    "    plt.tight_layout()\n",
    "    if dropmaxspecies == False:\n",
    "        plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ref_genome_hit_counts.png', bbox_inches='tight')\n",
    "    else:\n",
    "        plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ref_genome_hit_counts_nomaxspecies.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1521,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Generate the graphs\n",
    "generateGBGforRGHC(dropmaxspecies=False)\n",
    "generateGBGforRGHC(dropmaxspecies=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 805,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 839,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Changes rgblast pivot table count results to rgblast % within each barcode\n",
    "rgblast_joint_pivot_table_pc = rgblast_df_joint_pivot.copy()\n",
    "\n",
    "for spcs in rgblast_joint_pivot_table_pc.columns:\n",
    "    rgblast_joint_pivot_table_pc[str(spcs) + '%'] = np.round((rgblast_joint_pivot_table_pc[spcs] / rgblast_joint_pivot_table_pc['All'] * 100), 3)\n",
    "for spcs in rgblast_joint_pivot_table_pc.columns:\n",
    "    if '%' not in spcs:\n",
    "        rgblast_joint_pivot_table_pc.drop(spcs, axis=1, inplace=True)\n",
    "    \n",
    "rgblast_joint_pivot_table_pc.drop('All%', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1522,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Construct a grouped bar chart showing the reference genome hit % per barcode (log y axis)\n",
    "def generateGBGPCforRGHC(dropmaxspecies=False):\n",
    "    n_groups = len(rgblast_joint_pivot_table_pc.index.values)\n",
    "\n",
    "    max_genome = str(rgblast_joint_pivot_table_pc.ix['All'].idxmax(axis=1))\n",
    "\n",
    "    pos = list(range(len(rgblast_joint_pivot_table_pc[max_genome])))\n",
    "    width = 0.15\n",
    "    \n",
    "    columns_pc = list(rgblast_joint_pivot_table_pc.columns)\n",
    "    columns_nopc = list(map(lambda x:x.strip(\"%\"), columns_pc))\n",
    "    if dropmaxspecies == True:\n",
    "        columns_pc = [x for x in columns_pc if x != max_genome]\n",
    "        columns_nopc = [x for x in columns_nopc if x != max_genome.strip('%')]\n",
    "    \n",
    "    barcodes = list(rgblast_joint_pivot_table_pc.index)\n",
    "    fig, ax = plt.subplots(figsize=(8,3))\n",
    "\n",
    "    for ref_genome in columns_pc:  \n",
    "        plt.bar([p + width*int(columns_pc.index(ref_genome)) for p in pos], \n",
    "                rgblast_joint_pivot_table_pc[ref_genome], \n",
    "                width, \n",
    "                alpha=0.5, \n",
    "                color=random_color(x), \n",
    "                label=None)\n",
    "        \n",
    "    ax.set_ylabel('% of Reads', fontsize=14)\n",
    "    \n",
    "    rgblast_joint_pivot_table_pc_nomaxspecies = rgblast_joint_pivot_table_pc.copy()\n",
    "    rgblast_joint_pivot_table_pc_nomaxspecies.drop(max_genome, axis=1, inplace=True)\n",
    "    rgblast_joint_pivot_table_pc_nomaxspecies['All'] = rgblast_joint_pivot_table_pc_nomaxspecies.sum(axis=1)\n",
    "    \n",
    "    ax.set_ylim([0, int(rgblast_joint_pivot_table_pc_nomaxspecies['All'].values.max()+1)])\n",
    "    ax.set_yticks(range(0, int(rgblast_joint_pivot_table_pc_nomaxspecies['All'].values.max()+2), 1))\n",
    "    ax.set_xlabel('Barcode', fontsize=14)\n",
    "    ax.set_xticks([p + 2.5 * width for p in pos])\n",
    "    ax.set_xticklabels(barcodes)\n",
    "    ax.tick_params(top=\"off\", right=\"off\", bottom='off')\n",
    "    ax.set_title('Reads Hitting Reference Genomes - % Hit Per Genome By Barcode', y=1.08, fontweight='bold', fontsize=16)\n",
    "    \n",
    "    if dropmaxspecies == False:\n",
    "        for idx, bcs in enumerate(barcodes):\n",
    "            ax.annotate(str(float(\"{0:.2f}\".format(rgblast_joint_pivot_table_pc[max_genome][bcs]))), \n",
    "                     xy=(0.06, 0.95), xycoords='axes fraction', fontsize=10, xytext=(55*idx, 0), textcoords='offset points', \n",
    "                     ha='left', va='top')\n",
    "    \n",
    "    plt.xlim(min(pos) - width, max(pos) + 0.3 + width*4)\n",
    "    plt.legend(columns_nopc, loc='center left', bbox_to_anchor=(1.01, 0.5), fontsize=18, prop={'weight':'normal'})\n",
    "    plt.grid(b=None, which='major', axis='y', linestyle='dashed', linewidth=0.2)\n",
    "    if dropmaxspecies == False:\n",
    "        plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ref_genome_hit_percentage.png', bbox_inches='tight')\n",
    "    else:\n",
    "        plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ref_genome_hit_percentage_nomaxspecies.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1523,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "#Construct the graphs\n",
    "generateGBGPCforRGHC(dropmaxspecies=False)\n",
    "generateGBGPCforRGHC(dropmaxspecies=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 986,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Generate subsamples of your reference genome hits to be blasted\n",
    "os.chdir(work_dir + 'data/basecalled/')\n",
    "blast_hit_sample_id = \".blasthitsample_\"\n",
    "data_rghitsamples_path = work_dir + 'data/rghitsamples/'\n",
    "\n",
    "def generate_blasthit_samples(bcs, smp, rpts):\n",
    "    for rpt in range(1, rpts+1):\n",
    "        \n",
    "        rgblast_df_joint_sample = rgblast_df_joint.loc[rgblast_df_joint['barcode'] == str(bcs)].sample(n = smp)\n",
    "        rgblast_df_joint_sample_list = list(rgblast_df_joint_sample['qseqid'].values)\n",
    "        \n",
    "        rgblasthitsample_id = rgblast_file_id + blast_hit_sample_id + str(rpt) + fasta_file_type\n",
    "        rgblasthitsample_file = basecalled_fasta_file_id + bcs + rgblasthitsample_id\n",
    "\n",
    "        rgblasthitsample_sequences = []\n",
    "        if bcs == 'NB00':\n",
    "            for badbcs in bad_barcodes:\n",
    "                for seq in SeqIO.parse(open(basecalled_fasta_file_id + str(badbcs) + fasta_file_type), 'fasta'):\n",
    "                    if seq.id in rgblast_df_joint_sample_list:\n",
    "                        rgblasthitsample_sequences.append(seq)\n",
    "            SeqIO.write(rgblasthitsample_sequences, os.path.join(data_rghitsamples_path, \n",
    "                                                                 rgblasthitsample_file), \"fasta\")\n",
    "        else:\n",
    "            for seq in SeqIO.parse(open(basecalled_fasta_file_id + str(bcs) + fasta_file_type), 'fasta'):\n",
    "                if seq.id in rgblast_df_joint_sample_list:\n",
    "                    rgblasthitsample_sequences.append(seq)\n",
    "            SeqIO.write(rgblasthitsample_sequences, os.path.join(data_rghitsamples_path, \n",
    "                                                                 rgblasthitsample_file), \"fasta\")\n",
    "        \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 988,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Generate samples\n",
    "for bcs in good_barcodes:\n",
    "    generate_blasthit_samples(bcs, 1000, 3)\n",
    "generate_blasthit_samples('NB00', 500, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1454,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>pident</th>\n",
       "      <th>nident</th>\n",
       "      <th>sgi</th>\n",
       "      <th>sacc</th>\n",
       "      <th>staxids</th>\n",
       "      <th>sscinames</th>\n",
       "      <th>scomnames</th>\n",
       "      <th>sskingdoms</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fff48c4d-c4b4-476d-9bfe-5ef907f8af97_Basecall_...</td>\n",
       "      <td>gi|215104|gb|J02459.1|LAMCG</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>787.0</td>\n",
       "      <td>1272</td>\n",
       "      <td>79.95</td>\n",
       "      <td>1017</td>\n",
       "      <td>215104</td>\n",
       "      <td>J02459</td>\n",
       "      <td>10710</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4ac290f8-a386-40a5-ac85-2bf6e04bdf48_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2939.0</td>\n",
       "      <td>2581</td>\n",
       "      <td>88.30</td>\n",
       "      <td>2279</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4bcd6024-2a79-40fd-ab26-415daf5df51b_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1308.0</td>\n",
       "      <td>2463</td>\n",
       "      <td>78.36</td>\n",
       "      <td>1930</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4cdc928b-9ca4-403d-a69d-08c2cb471cc7_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1672.0</td>\n",
       "      <td>2200</td>\n",
       "      <td>82.09</td>\n",
       "      <td>1806</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4dbb7b75-bb28-4d97-8eea-99793ed8cdd5_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>9.000000e-126</td>\n",
       "      <td>462.0</td>\n",
       "      <td>1215</td>\n",
       "      <td>75.97</td>\n",
       "      <td>923</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4f87ce1b-e5c4-4642-8dc3-e1506c0386ec_Basecall_...</td>\n",
       "      <td>gi|215104|gb|J02459.1|LAMCG</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1663.0</td>\n",
       "      <td>2052</td>\n",
       "      <td>82.99</td>\n",
       "      <td>1703</td>\n",
       "      <td>215104</td>\n",
       "      <td>J02459</td>\n",
       "      <td>10710</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4fa5636f-14e0-49ea-b914-e6b667348e0c_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>736.0</td>\n",
       "      <td>819</td>\n",
       "      <td>84.25</td>\n",
       "      <td>690</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5004ce50-7801-4580-871c-f96e20111812_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1114.0</td>\n",
       "      <td>937</td>\n",
       "      <td>89.22</td>\n",
       "      <td>836</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>51ba06de-340a-4928-a642-cf1099776c34_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4060.0</td>\n",
       "      <td>3507</td>\n",
       "      <td>88.71</td>\n",
       "      <td>3111</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>51d404e6-945e-43fb-b990-ddc53addbb48_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2752.0</td>\n",
       "      <td>2831</td>\n",
       "      <td>85.66</td>\n",
       "      <td>2425</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>522694d6-1e70-4687-a282-cbaa0802b6b2_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>6.000000e-14</td>\n",
       "      <td>89.8</td>\n",
       "      <td>52</td>\n",
       "      <td>98.08</td>\n",
       "      <td>51</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>5232526b-3385-421a-a521-59988722ae46_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1962.0</td>\n",
       "      <td>2611</td>\n",
       "      <td>82.08</td>\n",
       "      <td>2143</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4a7e9d65-1bae-45c4-930c-673fe88ead84_Basecall_...</td>\n",
       "      <td>gi|857137982|gb|CP011495.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1413.0</td>\n",
       "      <td>1950</td>\n",
       "      <td>81.59</td>\n",
       "      <td>1591</td>\n",
       "      <td>857137982</td>\n",
       "      <td>CP011495</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>537e2da6-d349-42cc-9dbb-d54af2659cff_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2374.0</td>\n",
       "      <td>2031</td>\n",
       "      <td>89.07</td>\n",
       "      <td>1809</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>55a8b2cd-dd21-4af4-8bf4-b23bb8e3ad11_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>7.000000e-110</td>\n",
       "      <td>409.0</td>\n",
       "      <td>493</td>\n",
       "      <td>83.37</td>\n",
       "      <td>411</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>5734f1bc-a112-41e9-a4c8-2273702e85aa_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2396.0</td>\n",
       "      <td>2535</td>\n",
       "      <td>85.29</td>\n",
       "      <td>2162</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>57575801-172f-4b82-8af2-8f9f12f83e57_Basecall_...</td>\n",
       "      <td>gi|383396568|gb|JQ086373.1|</td>\n",
       "      <td>3.000000e-50</td>\n",
       "      <td>209.0</td>\n",
       "      <td>225</td>\n",
       "      <td>84.89</td>\n",
       "      <td>191</td>\n",
       "      <td>383396568</td>\n",
       "      <td>JQ086373</td>\n",
       "      <td>432200</td>\n",
       "      <td>Enterobacteria phage HK542</td>\n",
       "      <td>Enterobacteria phage HK542</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>58781adc-c25a-4c61-aa74-b92c8099428b_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1379</td>\n",
       "      <td>81.80</td>\n",
       "      <td>1128</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>59723d4d-ebe3-4e5d-8f20-1d50ef677112_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2351.0</td>\n",
       "      <td>2087</td>\n",
       "      <td>88.26</td>\n",
       "      <td>1842</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>59d93858-df94-47dc-b4ef-52655fc68fe4_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1775.0</td>\n",
       "      <td>1527</td>\n",
       "      <td>89.00</td>\n",
       "      <td>1359</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>5a03f229-3f7d-457c-b874-53f39de8554d_Basecall_...</td>\n",
       "      <td>gi|215104|gb|J02459.1|LAMCG</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3943.0</td>\n",
       "      <td>3499</td>\n",
       "      <td>88.23</td>\n",
       "      <td>3087</td>\n",
       "      <td>215104</td>\n",
       "      <td>J02459</td>\n",
       "      <td>10710</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>5a348401-0cc9-40b1-8feb-0579dc21fc0c_Basecall_...</td>\n",
       "      <td>gi|215104|gb|J02459.1|LAMCG</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2793.0</td>\n",
       "      <td>2181</td>\n",
       "      <td>90.88</td>\n",
       "      <td>1982</td>\n",
       "      <td>215104</td>\n",
       "      <td>J02459</td>\n",
       "      <td>10710</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>5ac8cd0e-3961-4adf-9949-25a7fe56a09c_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>1.000000e-19</td>\n",
       "      <td>108.0</td>\n",
       "      <td>65</td>\n",
       "      <td>96.92</td>\n",
       "      <td>63</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>5ba9ad0b-31d0-4b10-adea-917213ddfb29_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1777.0</td>\n",
       "      <td>1605</td>\n",
       "      <td>87.85</td>\n",
       "      <td>1410</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>5d150c4e-24f0-46d3-8bd8-aa81e8279fa7_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1325.0</td>\n",
       "      <td>1228</td>\n",
       "      <td>87.62</td>\n",
       "      <td>1076</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>54188dda-e250-4200-863a-2fb4618461aa_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3921.0</td>\n",
       "      <td>3472</td>\n",
       "      <td>88.31</td>\n",
       "      <td>3066</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>5dce57b3-f6b6-45fd-8d62-c4690d90d5fe_Basecall_...</td>\n",
       "      <td>gi|33334157|gb|AF547987.1|</td>\n",
       "      <td>2.000000e-72</td>\n",
       "      <td>283.0</td>\n",
       "      <td>307</td>\n",
       "      <td>84.69</td>\n",
       "      <td>260</td>\n",
       "      <td>33334157</td>\n",
       "      <td>AF547987</td>\n",
       "      <td>10761</td>\n",
       "      <td>Shigella phage Sf6</td>\n",
       "      <td>Shigella phage Sf6</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>4a5b41f0-bf9c-4d17-9732-f89dd4d957b4_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3964.0</td>\n",
       "      <td>3467</td>\n",
       "      <td>88.55</td>\n",
       "      <td>3070</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>49da0c5c-cadd-40dc-a0f5-cdadff067b58_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>9.000000e-122</td>\n",
       "      <td>448.0</td>\n",
       "      <td>508</td>\n",
       "      <td>83.86</td>\n",
       "      <td>426</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>40c08927-4017-4330-aa91-f5d88e120371_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4396.0</td>\n",
       "      <td>3513</td>\n",
       "      <td>90.32</td>\n",
       "      <td>3173</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BC01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22875</th>\n",
       "      <td>b694e6b0-be3f-41d4-a1ac-960beaa1e3cd_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1618.0</td>\n",
       "      <td>1512</td>\n",
       "      <td>87.37</td>\n",
       "      <td>1321</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22876</th>\n",
       "      <td>b1ef60cf-97c3-489f-961d-baf39ce220a1_Basecall_...</td>\n",
       "      <td>gi|445198867|gb|CP004045.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2442.0</td>\n",
       "      <td>2107</td>\n",
       "      <td>88.85</td>\n",
       "      <td>1872</td>\n",
       "      <td>445198867</td>\n",
       "      <td>CP004045</td>\n",
       "      <td>1282356</td>\n",
       "      <td>Pseudomonas poae RE*1-1-14</td>\n",
       "      <td>Pseudomonas poae RE*1-1-14</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22877</th>\n",
       "      <td>b0916066-2c1e-466c-9284-bb5743101691_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>885.0</td>\n",
       "      <td>803</td>\n",
       "      <td>87.80</td>\n",
       "      <td>705</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22878</th>\n",
       "      <td>b003f013-34ed-4b5b-a2f5-43092618a914_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1801.0</td>\n",
       "      <td>2051</td>\n",
       "      <td>84.01</td>\n",
       "      <td>1723</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22879</th>\n",
       "      <td>afd83f86-2fb3-4eb5-8ddc-ee3c5db3c0ab_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1803.0</td>\n",
       "      <td>1907</td>\n",
       "      <td>85.32</td>\n",
       "      <td>1627</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22880</th>\n",
       "      <td>ce8b74d8-dd50-434e-97b6-40f17d07aafc_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3181.0</td>\n",
       "      <td>2894</td>\n",
       "      <td>87.84</td>\n",
       "      <td>2542</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22881</th>\n",
       "      <td>f7695fba-01fa-4ca7-b6fb-2faa56221862_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1007.0</td>\n",
       "      <td>918</td>\n",
       "      <td>87.91</td>\n",
       "      <td>807</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22882</th>\n",
       "      <td>cf8bc4b8-bf79-48e8-9084-5bb9176bc532_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1557.0</td>\n",
       "      <td>1446</td>\n",
       "      <td>87.34</td>\n",
       "      <td>1263</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22883</th>\n",
       "      <td>d34c9a4d-3ab5-4a20-85de-10d5f02706b6_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1609.0</td>\n",
       "      <td>1945</td>\n",
       "      <td>83.24</td>\n",
       "      <td>1619</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22884</th>\n",
       "      <td>ffa2c16e-b7d6-4a8a-ae46-87afaa02881e_Basecall_...</td>\n",
       "      <td>gi|1002167280|gb|CP012733.1|</td>\n",
       "      <td>2.000000e-38</td>\n",
       "      <td>171.0</td>\n",
       "      <td>199</td>\n",
       "      <td>83.42</td>\n",
       "      <td>166</td>\n",
       "      <td>1002167280</td>\n",
       "      <td>CP012733</td>\n",
       "      <td>374923</td>\n",
       "      <td>Shigella flexneri 1a</td>\n",
       "      <td>Shigella flexneri 1a</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22885</th>\n",
       "      <td>fdb8137c-5ce3-407c-b99b-6ff38345ce2d_Basecall_...</td>\n",
       "      <td>gi|445198867|gb|CP004045.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2715.0</td>\n",
       "      <td>2632</td>\n",
       "      <td>86.63</td>\n",
       "      <td>2280</td>\n",
       "      <td>445198867</td>\n",
       "      <td>CP004045</td>\n",
       "      <td>1282356</td>\n",
       "      <td>Pseudomonas poae RE*1-1-14</td>\n",
       "      <td>Pseudomonas poae RE*1-1-14</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22886</th>\n",
       "      <td>f7e2b608-1731-4487-9ad2-18dee67a9e1f_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2036.0</td>\n",
       "      <td>2492</td>\n",
       "      <td>83.11</td>\n",
       "      <td>2071</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22887</th>\n",
       "      <td>f7799911-5a8f-46e7-b364-edba17e39acd_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>9.000000e-79</td>\n",
       "      <td>305.0</td>\n",
       "      <td>364</td>\n",
       "      <td>83.24</td>\n",
       "      <td>303</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22888</th>\n",
       "      <td>f5a788be-5f9a-46eb-9b60-e148a3733ddc_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>1159</td>\n",
       "      <td>84.12</td>\n",
       "      <td>975</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22889</th>\n",
       "      <td>f042288a-93df-406e-8000-d4c922156e54_Basecall_...</td>\n",
       "      <td>gi|215104|gb|J02459.1|LAMCG</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1122.0</td>\n",
       "      <td>2360</td>\n",
       "      <td>77.58</td>\n",
       "      <td>1831</td>\n",
       "      <td>215104</td>\n",
       "      <td>J02459</td>\n",
       "      <td>10710</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22890</th>\n",
       "      <td>ef48f553-5cec-43a2-9490-923f2b878b43_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1334.0</td>\n",
       "      <td>1601</td>\n",
       "      <td>83.45</td>\n",
       "      <td>1336</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22891</th>\n",
       "      <td>ee4935d3-5bcb-4ef2-b769-8eac9ecb7707_Basecall_...</td>\n",
       "      <td>gi|445198867|gb|CP004045.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>1827</td>\n",
       "      <td>87.79</td>\n",
       "      <td>1604</td>\n",
       "      <td>445198867</td>\n",
       "      <td>CP004045</td>\n",
       "      <td>1282356</td>\n",
       "      <td>Pseudomonas poae RE*1-1-14</td>\n",
       "      <td>Pseudomonas poae RE*1-1-14</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22892</th>\n",
       "      <td>e4a25176-ca1d-4d38-8270-f46670eb06c1_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>952.0</td>\n",
       "      <td>1743</td>\n",
       "      <td>78.43</td>\n",
       "      <td>1367</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22893</th>\n",
       "      <td>e1b86c65-8d8c-40e8-89fe-fd05c0f48b2d_Basecall_...</td>\n",
       "      <td>gi|628068230|ref|XM_007700579.1|</td>\n",
       "      <td>4.000000e-06</td>\n",
       "      <td>67.6</td>\n",
       "      <td>43</td>\n",
       "      <td>95.35</td>\n",
       "      <td>41</td>\n",
       "      <td>628068230</td>\n",
       "      <td>XM_007700579</td>\n",
       "      <td>665912</td>\n",
       "      <td>Bipolaris sorokiniana ND90Pr</td>\n",
       "      <td>Bipolaris sorokiniana ND90Pr</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22894</th>\n",
       "      <td>dd3809d4-4c13-4350-a3c8-a9238a4be437_Basecall_...</td>\n",
       "      <td>gi|215104|gb|J02459.1|LAMCG</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1616.0</td>\n",
       "      <td>1282</td>\n",
       "      <td>90.48</td>\n",
       "      <td>1160</td>\n",
       "      <td>215104</td>\n",
       "      <td>J02459</td>\n",
       "      <td>10710</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22895</th>\n",
       "      <td>db3dc26d-dc79-4f6d-bb32-078697042d1a_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>1600</td>\n",
       "      <td>80.38</td>\n",
       "      <td>1286</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22896</th>\n",
       "      <td>da417373-35f4-4d61-b72c-d06c2019e553_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1879.0</td>\n",
       "      <td>2833</td>\n",
       "      <td>80.59</td>\n",
       "      <td>2283</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22897</th>\n",
       "      <td>d9e9339b-243b-4c10-8895-720e67d62494_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>990.0</td>\n",
       "      <td>992</td>\n",
       "      <td>85.89</td>\n",
       "      <td>852</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22898</th>\n",
       "      <td>d83b594a-0033-4eba-ac7e-2e3ab156bc05_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2263.0</td>\n",
       "      <td>2816</td>\n",
       "      <td>82.74</td>\n",
       "      <td>2330</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22899</th>\n",
       "      <td>d7f99534-1d96-4d0a-80a7-e0e321d5c5af_Basecall_...</td>\n",
       "      <td>gi|215104|gb|J02459.1|LAMCG</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>706.0</td>\n",
       "      <td>620</td>\n",
       "      <td>88.55</td>\n",
       "      <td>549</td>\n",
       "      <td>215104</td>\n",
       "      <td>J02459</td>\n",
       "      <td>10710</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Enterobacteria phage lambda</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22900</th>\n",
       "      <td>d7465f86-7f13-485e-ad8d-2ccc505b6404_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>1.000000e-98</td>\n",
       "      <td>372.0</td>\n",
       "      <td>590</td>\n",
       "      <td>79.66</td>\n",
       "      <td>470</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22901</th>\n",
       "      <td>d735aed3-00de-4486-9816-5305c651b7d3_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>985.0</td>\n",
       "      <td>1057</td>\n",
       "      <td>85.15</td>\n",
       "      <td>900</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22902</th>\n",
       "      <td>d5a625b0-eaa2-4fe0-9b77-bce668868a07_Basecall_...</td>\n",
       "      <td>gi|1008889473|gb|CP014768.1|</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>662.0</td>\n",
       "      <td>838</td>\n",
       "      <td>82.70</td>\n",
       "      <td>693</td>\n",
       "      <td>1008889473</td>\n",
       "      <td>CP014768</td>\n",
       "      <td>1813821</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Shigella sp. PAMC 28760</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22903</th>\n",
       "      <td>d1030503-8842-4cdd-83c6-7b42af9f13be_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3866.0</td>\n",
       "      <td>3509</td>\n",
       "      <td>87.92</td>\n",
       "      <td>3085</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22904</th>\n",
       "      <td>02014d7e-c19d-4968-b211-b7c4b79dd04c_Basecall_...</td>\n",
       "      <td>gi|1051181|gb|U37692.1|CVU37692</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2082.0</td>\n",
       "      <td>2941</td>\n",
       "      <td>81.33</td>\n",
       "      <td>2392</td>\n",
       "      <td>1051181</td>\n",
       "      <td>U37692</td>\n",
       "      <td>43300</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>Cloning vector lambda TXF97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NB00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22905 rows  14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  qseqid  \\\n",
       "0      fff48c4d-c4b4-476d-9bfe-5ef907f8af97_Basecall_...   \n",
       "1      4ac290f8-a386-40a5-ac85-2bf6e04bdf48_Basecall_...   \n",
       "2      4bcd6024-2a79-40fd-ab26-415daf5df51b_Basecall_...   \n",
       "3      4cdc928b-9ca4-403d-a69d-08c2cb471cc7_Basecall_...   \n",
       "4      4dbb7b75-bb28-4d97-8eea-99793ed8cdd5_Basecall_...   \n",
       "5      4f87ce1b-e5c4-4642-8dc3-e1506c0386ec_Basecall_...   \n",
       "6      4fa5636f-14e0-49ea-b914-e6b667348e0c_Basecall_...   \n",
       "7      5004ce50-7801-4580-871c-f96e20111812_Basecall_...   \n",
       "8      51ba06de-340a-4928-a642-cf1099776c34_Basecall_...   \n",
       "9      51d404e6-945e-43fb-b990-ddc53addbb48_Basecall_...   \n",
       "10     522694d6-1e70-4687-a282-cbaa0802b6b2_Basecall_...   \n",
       "11     5232526b-3385-421a-a521-59988722ae46_Basecall_...   \n",
       "12     4a7e9d65-1bae-45c4-930c-673fe88ead84_Basecall_...   \n",
       "13     537e2da6-d349-42cc-9dbb-d54af2659cff_Basecall_...   \n",
       "14     55a8b2cd-dd21-4af4-8bf4-b23bb8e3ad11_Basecall_...   \n",
       "15     5734f1bc-a112-41e9-a4c8-2273702e85aa_Basecall_...   \n",
       "16     57575801-172f-4b82-8af2-8f9f12f83e57_Basecall_...   \n",
       "17     58781adc-c25a-4c61-aa74-b92c8099428b_Basecall_...   \n",
       "18     59723d4d-ebe3-4e5d-8f20-1d50ef677112_Basecall_...   \n",
       "19     59d93858-df94-47dc-b4ef-52655fc68fe4_Basecall_...   \n",
       "20     5a03f229-3f7d-457c-b874-53f39de8554d_Basecall_...   \n",
       "21     5a348401-0cc9-40b1-8feb-0579dc21fc0c_Basecall_...   \n",
       "22     5ac8cd0e-3961-4adf-9949-25a7fe56a09c_Basecall_...   \n",
       "23     5ba9ad0b-31d0-4b10-adea-917213ddfb29_Basecall_...   \n",
       "24     5d150c4e-24f0-46d3-8bd8-aa81e8279fa7_Basecall_...   \n",
       "25     54188dda-e250-4200-863a-2fb4618461aa_Basecall_...   \n",
       "26     5dce57b3-f6b6-45fd-8d62-c4690d90d5fe_Basecall_...   \n",
       "27     4a5b41f0-bf9c-4d17-9732-f89dd4d957b4_Basecall_...   \n",
       "28     49da0c5c-cadd-40dc-a0f5-cdadff067b58_Basecall_...   \n",
       "29     40c08927-4017-4330-aa91-f5d88e120371_Basecall_...   \n",
       "...                                                  ...   \n",
       "22875  b694e6b0-be3f-41d4-a1ac-960beaa1e3cd_Basecall_...   \n",
       "22876  b1ef60cf-97c3-489f-961d-baf39ce220a1_Basecall_...   \n",
       "22877  b0916066-2c1e-466c-9284-bb5743101691_Basecall_...   \n",
       "22878  b003f013-34ed-4b5b-a2f5-43092618a914_Basecall_...   \n",
       "22879  afd83f86-2fb3-4eb5-8ddc-ee3c5db3c0ab_Basecall_...   \n",
       "22880  ce8b74d8-dd50-434e-97b6-40f17d07aafc_Basecall_...   \n",
       "22881  f7695fba-01fa-4ca7-b6fb-2faa56221862_Basecall_...   \n",
       "22882  cf8bc4b8-bf79-48e8-9084-5bb9176bc532_Basecall_...   \n",
       "22883  d34c9a4d-3ab5-4a20-85de-10d5f02706b6_Basecall_...   \n",
       "22884  ffa2c16e-b7d6-4a8a-ae46-87afaa02881e_Basecall_...   \n",
       "22885  fdb8137c-5ce3-407c-b99b-6ff38345ce2d_Basecall_...   \n",
       "22886  f7e2b608-1731-4487-9ad2-18dee67a9e1f_Basecall_...   \n",
       "22887  f7799911-5a8f-46e7-b364-edba17e39acd_Basecall_...   \n",
       "22888  f5a788be-5f9a-46eb-9b60-e148a3733ddc_Basecall_...   \n",
       "22889  f042288a-93df-406e-8000-d4c922156e54_Basecall_...   \n",
       "22890  ef48f553-5cec-43a2-9490-923f2b878b43_Basecall_...   \n",
       "22891  ee4935d3-5bcb-4ef2-b769-8eac9ecb7707_Basecall_...   \n",
       "22892  e4a25176-ca1d-4d38-8270-f46670eb06c1_Basecall_...   \n",
       "22893  e1b86c65-8d8c-40e8-89fe-fd05c0f48b2d_Basecall_...   \n",
       "22894  dd3809d4-4c13-4350-a3c8-a9238a4be437_Basecall_...   \n",
       "22895  db3dc26d-dc79-4f6d-bb32-078697042d1a_Basecall_...   \n",
       "22896  da417373-35f4-4d61-b72c-d06c2019e553_Basecall_...   \n",
       "22897  d9e9339b-243b-4c10-8895-720e67d62494_Basecall_...   \n",
       "22898  d83b594a-0033-4eba-ac7e-2e3ab156bc05_Basecall_...   \n",
       "22899  d7f99534-1d96-4d0a-80a7-e0e321d5c5af_Basecall_...   \n",
       "22900  d7465f86-7f13-485e-ad8d-2ccc505b6404_Basecall_...   \n",
       "22901  d735aed3-00de-4486-9816-5305c651b7d3_Basecall_...   \n",
       "22902  d5a625b0-eaa2-4fe0-9b77-bce668868a07_Basecall_...   \n",
       "22903  d1030503-8842-4cdd-83c6-7b42af9f13be_Basecall_...   \n",
       "22904  02014d7e-c19d-4968-b211-b7c4b79dd04c_Basecall_...   \n",
       "\n",
       "                                 sseqid         evalue  bitscore  length  \\\n",
       "0           gi|215104|gb|J02459.1|LAMCG   0.000000e+00     787.0    1272   \n",
       "1       gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    2939.0    2581   \n",
       "2       gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1308.0    2463   \n",
       "3       gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1672.0    2200   \n",
       "4          gi|1008889473|gb|CP014768.1|  9.000000e-126     462.0    1215   \n",
       "5           gi|215104|gb|J02459.1|LAMCG   0.000000e+00    1663.0    2052   \n",
       "6          gi|1008889473|gb|CP014768.1|   0.000000e+00     736.0     819   \n",
       "7       gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1114.0     937   \n",
       "8       gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    4060.0    3507   \n",
       "9       gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    2752.0    2831   \n",
       "10         gi|1008889473|gb|CP014768.1|   6.000000e-14      89.8      52   \n",
       "11      gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1962.0    2611   \n",
       "12          gi|857137982|gb|CP011495.1|   0.000000e+00    1413.0    1950   \n",
       "13      gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    2374.0    2031   \n",
       "14         gi|1008889473|gb|CP014768.1|  7.000000e-110     409.0     493   \n",
       "15      gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    2396.0    2535   \n",
       "16          gi|383396568|gb|JQ086373.1|   3.000000e-50     209.0     225   \n",
       "17         gi|1008889473|gb|CP014768.1|   0.000000e+00    1000.0    1379   \n",
       "18      gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    2351.0    2087   \n",
       "19         gi|1008889473|gb|CP014768.1|   0.000000e+00    1775.0    1527   \n",
       "20          gi|215104|gb|J02459.1|LAMCG   0.000000e+00    3943.0    3499   \n",
       "21          gi|215104|gb|J02459.1|LAMCG   0.000000e+00    2793.0    2181   \n",
       "22         gi|1008889473|gb|CP014768.1|   1.000000e-19     108.0      65   \n",
       "23      gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1777.0    1605   \n",
       "24      gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1325.0    1228   \n",
       "25      gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    3921.0    3472   \n",
       "26           gi|33334157|gb|AF547987.1|   2.000000e-72     283.0     307   \n",
       "27      gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    3964.0    3467   \n",
       "28         gi|1008889473|gb|CP014768.1|  9.000000e-122     448.0     508   \n",
       "29      gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    4396.0    3513   \n",
       "...                                 ...            ...       ...     ...   \n",
       "22875      gi|1008889473|gb|CP014768.1|   0.000000e+00    1618.0    1512   \n",
       "22876       gi|445198867|gb|CP004045.1|   0.000000e+00    2442.0    2107   \n",
       "22877   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00     885.0     803   \n",
       "22878   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1801.0    2051   \n",
       "22879      gi|1008889473|gb|CP014768.1|   0.000000e+00    1803.0    1907   \n",
       "22880      gi|1008889473|gb|CP014768.1|   0.000000e+00    3181.0    2894   \n",
       "22881   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1007.0     918   \n",
       "22882      gi|1008889473|gb|CP014768.1|   0.000000e+00    1557.0    1446   \n",
       "22883   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1609.0    1945   \n",
       "22884      gi|1002167280|gb|CP012733.1|   2.000000e-38     171.0     199   \n",
       "22885       gi|445198867|gb|CP004045.1|   0.000000e+00    2715.0    2632   \n",
       "22886   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    2036.0    2492   \n",
       "22887      gi|1008889473|gb|CP014768.1|   9.000000e-79     305.0     364   \n",
       "22888      gi|1008889473|gb|CP014768.1|   0.000000e+00    1027.0    1159   \n",
       "22889       gi|215104|gb|J02459.1|LAMCG   0.000000e+00    1122.0    2360   \n",
       "22890      gi|1008889473|gb|CP014768.1|   0.000000e+00    1334.0    1601   \n",
       "22891       gi|445198867|gb|CP004045.1|   0.000000e+00    2013.0    1827   \n",
       "22892      gi|1008889473|gb|CP014768.1|   0.000000e+00     952.0    1743   \n",
       "22893  gi|628068230|ref|XM_007700579.1|   4.000000e-06      67.6      43   \n",
       "22894       gi|215104|gb|J02459.1|LAMCG   0.000000e+00    1616.0    1282   \n",
       "22895      gi|1008889473|gb|CP014768.1|   0.000000e+00    1057.0    1600   \n",
       "22896   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    1879.0    2833   \n",
       "22897   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00     990.0     992   \n",
       "22898   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    2263.0    2816   \n",
       "22899       gi|215104|gb|J02459.1|LAMCG   0.000000e+00     706.0     620   \n",
       "22900      gi|1008889473|gb|CP014768.1|   1.000000e-98     372.0     590   \n",
       "22901   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00     985.0    1057   \n",
       "22902      gi|1008889473|gb|CP014768.1|   0.000000e+00     662.0     838   \n",
       "22903   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    3866.0    3509   \n",
       "22904   gi|1051181|gb|U37692.1|CVU37692   0.000000e+00    2082.0    2941   \n",
       "\n",
       "       pident  nident         sgi          sacc  staxids  \\\n",
       "0       79.95    1017      215104        J02459    10710   \n",
       "1       88.30    2279     1051181        U37692    43300   \n",
       "2       78.36    1930     1051181        U37692    43300   \n",
       "3       82.09    1806     1051181        U37692    43300   \n",
       "4       75.97     923  1008889473      CP014768  1813821   \n",
       "5       82.99    1703      215104        J02459    10710   \n",
       "6       84.25     690  1008889473      CP014768  1813821   \n",
       "7       89.22     836     1051181        U37692    43300   \n",
       "8       88.71    3111     1051181        U37692    43300   \n",
       "9       85.66    2425     1051181        U37692    43300   \n",
       "10      98.08      51  1008889473      CP014768  1813821   \n",
       "11      82.08    2143     1051181        U37692    43300   \n",
       "12      81.59    1591   857137982      CP011495      562   \n",
       "13      89.07    1809     1051181        U37692    43300   \n",
       "14      83.37     411  1008889473      CP014768  1813821   \n",
       "15      85.29    2162     1051181        U37692    43300   \n",
       "16      84.89     191   383396568      JQ086373   432200   \n",
       "17      81.80    1128  1008889473      CP014768  1813821   \n",
       "18      88.26    1842     1051181        U37692    43300   \n",
       "19      89.00    1359  1008889473      CP014768  1813821   \n",
       "20      88.23    3087      215104        J02459    10710   \n",
       "21      90.88    1982      215104        J02459    10710   \n",
       "22      96.92      63  1008889473      CP014768  1813821   \n",
       "23      87.85    1410     1051181        U37692    43300   \n",
       "24      87.62    1076     1051181        U37692    43300   \n",
       "25      88.31    3066     1051181        U37692    43300   \n",
       "26      84.69     260    33334157      AF547987    10761   \n",
       "27      88.55    3070     1051181        U37692    43300   \n",
       "28      83.86     426  1008889473      CP014768  1813821   \n",
       "29      90.32    3173     1051181        U37692    43300   \n",
       "...       ...     ...         ...           ...      ...   \n",
       "22875   87.37    1321  1008889473      CP014768  1813821   \n",
       "22876   88.85    1872   445198867      CP004045  1282356   \n",
       "22877   87.80     705     1051181        U37692    43300   \n",
       "22878   84.01    1723     1051181        U37692    43300   \n",
       "22879   85.32    1627  1008889473      CP014768  1813821   \n",
       "22880   87.84    2542  1008889473      CP014768  1813821   \n",
       "22881   87.91     807     1051181        U37692    43300   \n",
       "22882   87.34    1263  1008889473      CP014768  1813821   \n",
       "22883   83.24    1619     1051181        U37692    43300   \n",
       "22884   83.42     166  1002167280      CP012733   374923   \n",
       "22885   86.63    2280   445198867      CP004045  1282356   \n",
       "22886   83.11    2071     1051181        U37692    43300   \n",
       "22887   83.24     303  1008889473      CP014768  1813821   \n",
       "22888   84.12     975  1008889473      CP014768  1813821   \n",
       "22889   77.58    1831      215104        J02459    10710   \n",
       "22890   83.45    1336  1008889473      CP014768  1813821   \n",
       "22891   87.79    1604   445198867      CP004045  1282356   \n",
       "22892   78.43    1367  1008889473      CP014768  1813821   \n",
       "22893   95.35      41   628068230  XM_007700579   665912   \n",
       "22894   90.48    1160      215104        J02459    10710   \n",
       "22895   80.38    1286  1008889473      CP014768  1813821   \n",
       "22896   80.59    2283     1051181        U37692    43300   \n",
       "22897   85.89     852     1051181        U37692    43300   \n",
       "22898   82.74    2330     1051181        U37692    43300   \n",
       "22899   88.55     549      215104        J02459    10710   \n",
       "22900   79.66     470  1008889473      CP014768  1813821   \n",
       "22901   85.15     900     1051181        U37692    43300   \n",
       "22902   82.70     693  1008889473      CP014768  1813821   \n",
       "22903   87.92    3085     1051181        U37692    43300   \n",
       "22904   81.33    2392     1051181        U37692    43300   \n",
       "\n",
       "                          sscinames                     scomnames sskingdoms  \\\n",
       "0       Enterobacteria phage lambda   Enterobacteria phage lambda    Viruses   \n",
       "1       Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "2       Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "3       Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "4           Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "5       Enterobacteria phage lambda   Enterobacteria phage lambda    Viruses   \n",
       "6           Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "7       Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "8       Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "9       Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "10          Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "11      Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "12                 Escherichia coli              Escherichia coli   Bacteria   \n",
       "13      Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "14          Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "15      Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "16       Enterobacteria phage HK542    Enterobacteria phage HK542    Viruses   \n",
       "17          Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "18      Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "19          Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "20      Enterobacteria phage lambda   Enterobacteria phage lambda    Viruses   \n",
       "21      Enterobacteria phage lambda   Enterobacteria phage lambda    Viruses   \n",
       "22          Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "23      Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "24      Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "25      Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "26               Shigella phage Sf6            Shigella phage Sf6    Viruses   \n",
       "27      Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "28          Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "29      Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "...                             ...                           ...        ...   \n",
       "22875       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22876    Pseudomonas poae RE*1-1-14    Pseudomonas poae RE*1-1-14   Bacteria   \n",
       "22877   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22878   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22879       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22880       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22881   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22882       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22883   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22884          Shigella flexneri 1a          Shigella flexneri 1a   Bacteria   \n",
       "22885    Pseudomonas poae RE*1-1-14    Pseudomonas poae RE*1-1-14   Bacteria   \n",
       "22886   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22887       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22888       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22889   Enterobacteria phage lambda   Enterobacteria phage lambda    Viruses   \n",
       "22890       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22891    Pseudomonas poae RE*1-1-14    Pseudomonas poae RE*1-1-14   Bacteria   \n",
       "22892       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22893  Bipolaris sorokiniana ND90Pr  Bipolaris sorokiniana ND90Pr  Eukaryota   \n",
       "22894   Enterobacteria phage lambda   Enterobacteria phage lambda    Viruses   \n",
       "22895       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22896   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22897   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22898   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22899   Enterobacteria phage lambda   Enterobacteria phage lambda    Viruses   \n",
       "22900       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22901   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22902       Shigella sp. PAMC 28760       Shigella sp. PAMC 28760   Bacteria   \n",
       "22903   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "22904   Cloning vector lambda TXF97   Cloning vector lambda TXF97        NaN   \n",
       "\n",
       "      barcode  \n",
       "0        BC01  \n",
       "1        BC01  \n",
       "2        BC01  \n",
       "3        BC01  \n",
       "4        BC01  \n",
       "5        BC01  \n",
       "6        BC01  \n",
       "7        BC01  \n",
       "8        BC01  \n",
       "9        BC01  \n",
       "10       BC01  \n",
       "11       BC01  \n",
       "12       BC01  \n",
       "13       BC01  \n",
       "14       BC01  \n",
       "15       BC01  \n",
       "16       BC01  \n",
       "17       BC01  \n",
       "18       BC01  \n",
       "19       BC01  \n",
       "20       BC01  \n",
       "21       BC01  \n",
       "22       BC01  \n",
       "23       BC01  \n",
       "24       BC01  \n",
       "25       BC01  \n",
       "26       BC01  \n",
       "27       BC01  \n",
       "28       BC01  \n",
       "29       BC01  \n",
       "...       ...  \n",
       "22875    NB00  \n",
       "22876    NB00  \n",
       "22877    NB00  \n",
       "22878    NB00  \n",
       "22879    NB00  \n",
       "22880    NB00  \n",
       "22881    NB00  \n",
       "22882    NB00  \n",
       "22883    NB00  \n",
       "22884    NB00  \n",
       "22885    NB00  \n",
       "22886    NB00  \n",
       "22887    NB00  \n",
       "22888    NB00  \n",
       "22889    NB00  \n",
       "22890    NB00  \n",
       "22891    NB00  \n",
       "22892    NB00  \n",
       "22893    NB00  \n",
       "22894    NB00  \n",
       "22895    NB00  \n",
       "22896    NB00  \n",
       "22897    NB00  \n",
       "22898    NB00  \n",
       "22899    NB00  \n",
       "22900    NB00  \n",
       "22901    NB00  \n",
       "22902    NB00  \n",
       "22903    NB00  \n",
       "22904    NB00  \n",
       "\n",
       "[22905 rows x 14 columns]"
      ]
     },
     "execution_count": 1454,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1099,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#def generate_notblasted_counts(bcs):\n",
    "    #read_id_list = ''\n",
    "    #read_id_list = basecalled_df.groupby('Barcode')['Read_id'].get_group(bcs).tolist()\n",
    "    #qseqid_list = ''\n",
    "    #qseqid_list = rgblast_df_joint.groupby('barcode')['qseqid'].get_group(bcs).tolist()\n",
    "    #not_rgblasted_list = sorted(list(set(read_id_list) - set(qseqid_list)))\n",
    "    #print(len(read_id_list))\n",
    "    #print(len(qseqid_list))\n",
    "    #print(len(not_rgblasted_list))\n",
    "    \n",
    "#generate_notblasted_counts('BC03')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1434,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Generates fasta files with/without the reads hitting the reference genomes, \n",
    "#sorts them to their appropriate folder location\n",
    "os.chdir(work_dir + 'data/basecalled/')\n",
    "def generate_rghit_fasta(bcs, rghit):\n",
    "    \"\"\"Takes basecalled.df, and list of read_ids from rgblast_df_joint. For a single barcode:\n",
    "    - if rghit is yes, extract a fasta file with all the reads hitting the reference genomes to data/rghityes\n",
    "    - if rghit is no, extract a fasta file with all the reads not hitting the reference genomes to data/rghitno\"\"\"\n",
    "    os.chdir(work_dir + 'data/basecalled/')\n",
    "    read_id_list = basecalled_df.groupby('Barcode')['Read_id'].get_group(bcs).tolist()\n",
    "    qseqid_list = rgblast_df_joint.groupby('barcode')['qseqid'].get_group(bcs).tolist()\n",
    "\n",
    "    if rghit == 'Yes':\n",
    "        rgblasted_hit_id = '.blasthit'\n",
    "        rgblasthit_id = rgblast_file_id + rgblasted_hit_id + fasta_file_type\n",
    "        rgblasthit_file = basecalled_fasta_file_id + bcs + rgblasthit_id\n",
    "        \n",
    "        if bcs == 'NB00':\n",
    "            rghityes = []\n",
    "            for badbcs in bad_barcodes:\n",
    "                rghityes_NB00 = ''\n",
    "                rghityes_NB00 = (rec for rec in SeqIO.parse(basecalled_fasta_file_id + str(badbcs) + fasta_file_type, 'fasta') if rec.id in qseqid_list) \n",
    "                rghityes.extend(rghityes_NB00)\n",
    "            count = ''\n",
    "            count = SeqIO.write(rghityes, os.path.join(work_dir + 'data/rghityes/', rgblasthit_file), \"fasta\")\n",
    "        else:\n",
    "            rghityes = ''\n",
    "            rghityes = (rec for rec in SeqIO.parse(basecalled_fasta_file_id + str(bcs) + fasta_file_type, 'fasta') if rec.id in qseqid_list) \n",
    "            count = ''\n",
    "            count = SeqIO.write(rghityes, os.path.join(work_dir + 'data/rghityes/', rgblasthit_file), \"fasta\")\n",
    "            \n",
    "    if rghit == 'No':   \n",
    "        norgblasted_hit_id = '.noblasthit'\n",
    "        norgblasthit_id = rgblast_file_id + norgblasted_hit_id + fasta_file_type\n",
    "        norgblasthit_file = basecalled_fasta_file_id + bcs + norgblasthit_id\n",
    "        \n",
    "        if bcs == 'NB00':\n",
    "            rghitno = []\n",
    "            for badbcs in bad_barcodes:\n",
    "                rghitno_NB00 = ''\n",
    "                rghitno_NB00 = (rec for rec in SeqIO.parse(basecalled_fasta_file_id + str(badbcs) + fasta_file_type, 'fasta') if rec.id not in qseqid_list) \n",
    "                rghitno.extend(rghitno_NB00)\n",
    "            count = ''\n",
    "            count = SeqIO.write(rghitno, os.path.join(work_dir + 'data/rghitno/', norgblasthit_file), \"fasta\")\n",
    "        else:\n",
    "            rghitno = ''\n",
    "            rghitno = (rec for rec in SeqIO.parse(basecalled_fasta_file_id + str(bcs) + fasta_file_type, 'fasta') if rec.id not in qseqid_list) \n",
    "            count = ''\n",
    "            count = SeqIO.write(rghitno, os.path.join(work_dir + 'data/rghitno/', norgblasthit_file), \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1062,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved 3294 records\n"
     ]
    }
   ],
   "source": [
    "for bcs in list(rgblast_df_joint.barcode.unique()):\n",
    "    generate_rghit_fasta(bcs, 'Yes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1431,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for bcs in list(rgblast_df_joint.barcode.unique()):\n",
    "    generate_rghit_fasta(bcs, 'No')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1514,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#create a table summarising blasted counts, hit vs no hit counts and %\n",
    "rgblasted_count_df = pd.concat([basecalled_df_pivot['len'], rgblast_df_joint_pivot['All']],  axis=1)\n",
    "rgblasted_count_df.applymap(lambda x: int(x))\n",
    "rgblasted_count_df.rename(columns={'len' : 'RGCount', 'All' : 'RGHitYes'}, inplace=True)\n",
    "rgblasted_count_df['RGHitNo'] = rgblasted_count_df['RGCount'] - rgblasted_count_df['RGHitYes']\n",
    "rgblasted_count_df['RGHitYes%'] = np.round((rgblasted_count_df['RGHitYes'] / rgblasted_count_df['RGCount'] * 100), 3)\n",
    "rgblasted_count_df['RGHitNo%'] = np.round((rgblasted_count_df['RGHitNo'] / rgblasted_count_df['RGCount'] * 100), 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1515,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "rgblasted_count_df_noall = rgblasted_count_df.drop('All')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1516,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Saves pivot table summary in analysis/tables/ folder ###\n",
    "fig = plt.figure(figsize=(10,2))\n",
    "ax = fig.add_subplot(1, 1, 1, frame_on=False)\n",
    "ax.xaxis.set_visible(False)  # hide the x axis\n",
    "ax.yaxis.set_visible(False)  # hide the y axis\n",
    "\n",
    "table(ax, rgblasted_count_df, loc='center')\n",
    "\n",
    "plt.savefig(os.path.join(work_dir + 'analysis/tables/', \n",
    "                        basecalled_fasta_file_id + \"rgblast_hit_vs_nohit_summary.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1524,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "#two qualitative variables - barcode, blasted vs no hit vs hit\n",
    "#one quantitative - read count\n",
    "n_groups = len(rgblasted_count_df_noall.index.values)\n",
    "\n",
    "bar_labels = ['Blasted', 'Blasted and Hit', 'Blasted and No Hit']\n",
    "counts = [tuple(rgblasted_count_df_noall['RGCount']),\n",
    "          tuple(rgblasted_count_df_noall['RGHitYes']),\n",
    "          tuple(rgblasted_count_df_noall['RGHitNo'])]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6.5,3))\n",
    "\n",
    "index = np.arange(n_groups)\n",
    "bar_width = 0.15\n",
    "opacity = 0.4\n",
    "\n",
    "for idx, (lbl, cnts) in enumerate(zip(bar_labels, counts)):\n",
    "    plt.bar(index + bar_width*idx, cnts, bar_width,\n",
    "                 alpha=opacity,\n",
    "                 color=random_color(x),\n",
    "                 label=lbl)\n",
    "    \n",
    "plt.xlabel('Barcode')\n",
    "plt.ylabel('Read Count')\n",
    "plt.title('Number of Reads Blasted and Hit To Reference Genomes', y=1.08)\n",
    "plt.xticks(index + bar_width*1.5, tuple(rgblasted_count_df_noall.index.values))\n",
    "plt.tick_params(right='off', top='off', bottom='off')\n",
    "plt.legend(loc='center left', bbox_to_anchor=(1.01, 0.5), fontsize=10)\n",
    "plt.grid(b=None, which='major', axis='y', linestyle='dashed', linewidth=0.2)\n",
    "plt.margins(x=0.1)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ref_genome_hit_and_nohit_counts.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1525,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "#two qualitative variables - barcode, no hit vs hit\n",
    "#one quantitative - read count\n",
    "n_groups = len(rgblasted_count_df_noall.index.values)\n",
    "\n",
    "bar_labels = ['Blasted and Hit', 'Blasted and No Hit']\n",
    "counts = [tuple(rgblasted_count_df_noall['RGHitYes%']),\n",
    "          tuple(rgblasted_count_df_noall['RGHitNo%'])]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6.5,3))\n",
    "\n",
    "index = np.arange(n_groups)\n",
    "bar_width = 0.15\n",
    "opacity = 0.4\n",
    "\n",
    "for idx, (lbl, cnts) in enumerate(zip(bar_labels, counts)):\n",
    "    plt.bar(index + bar_width*idx, cnts, bar_width,\n",
    "                 alpha=opacity,\n",
    "                 color=random_color(x),\n",
    "                 label=lbl)\n",
    "\n",
    "plt.xlabel('Barcode')\n",
    "plt.ylabel('% of Reads Per Barcode')\n",
    "plt.title('% of Reads Blasted and Hit To Reference Genomes', y=1.08)\n",
    "plt.xticks(index + bar_width, tuple(rgblasted_count_df_noall.index.values), y=-0.01)\n",
    "plt.tick_params(top='off', right='off')\n",
    "plt.legend(loc='center left', bbox_to_anchor=(1.01, 0.5), fontsize=12)\n",
    "plt.grid(b=None, which='major', axis='y', linestyle='dashed', linewidth=0.2)\n",
    "plt.margins(x=0.1)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ref_genome_hit_and_nohit_percentage.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1152,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Generate taxonomic database from ete3\n",
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1608,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Extracts ncbiblast files from the initial input folder and moves them to data/basecalled to be manipulated\n",
    "ncbiblast_input_file_names = os.listdir(ncbiblast_input_dir_path)\n",
    "ncbiblast_input_file_names.sort()\n",
    "if ncbiblast_input_file_names:\n",
    "    ncbiblast_file_type = get_file_type(ncbiblast_input_file_names[-1], '.')\n",
    "\n",
    "    for ncbiblast_file in ncbiblast_input_file_names:\n",
    "        if os.path.exists(ncbiblast_input_dir_path + ncbiblast_file):\n",
    "            if not os.path.exists(work_dir + 'blast/ncbibesthit/' + ncbiblast_file):\n",
    "                os.rename(ncbiblast_input_dir_path + ncbiblast_file, work_dir + 'blast/ncbibesthit/' + ncbiblast_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1609,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Gets a list of all the ncbiblast hit file names from .blast/ncbibesthit\n",
    "# creates a list of all the reference genome blast besthit file names\n",
    "#Generates a list of all the fasta files present in data/basecalled\n",
    "ncbiblast_file_names = os.listdir(work_dir + 'blast/ncbibesthit')\n",
    "ncbiblast_besthit_file_names = [x for x in ncbiblast_file_names if ncbiblast_file_type in x]\n",
    "ncbiblast_besthit_file_names.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1610,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Extracts the ncbi blast id from the ncbiblast hit file names\n",
    "ncbiblast_file_id_list = []\n",
    "for file_name in ncbiblast_besthit_file_names:\n",
    "    start = file_name.find(fasta_file_type)\n",
    "    end = file_name.rfind(ncbiblast_file_type)\n",
    "    ncbiblast_file_id_list.append(file_name[start+len(fasta_file_type):end])\n",
    "ncbiblast_file_id = list(set(ncbiblast_file_id_list))[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1639,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Make a list of all available barcodes the ncbibesthit reads are sorted into, based on ncbibesthit file names\n",
    "ncbiblast_besthit_file_names_idstrip = [file_name.strip(basecalled_fasta_file_id) for file_name in ncbiblast_besthit_file_names]\n",
    "ncbiblast_besthit_barcodes = []\n",
    "for bcs in range(0, len(ncbiblast_besthit_file_names_idstrip)):\n",
    "    bclabel_index = ''\n",
    "    if int(ncbiblast_besthit_file_names_idstrip[bcs].find('BC')) > int(ncbiblast_besthit_file_names_idstrip[bcs].find('NB')):\n",
    "        bclabel_index = int(ncbiblast_besthit_file_names_idstrip[bcs].find('BC'))\n",
    "    else:\n",
    "        bclabel_index = int(ncbiblast_besthit_file_names_idstrip[bcs].find('NB'))\n",
    "    ncbiblast_besthit_barcodes.append(ncbiblast_besthit_file_names_idstrip[bcs][bclabel_index:bclabel_index + 4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1640,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Reads each ncbiblast_besthit file as a dataframe, appends a column with the relevant barcode\n",
    "os.chdir(work_dir + 'blast/ncbibesthit')\n",
    "ncbiblast_dfs_dict = {}\n",
    "blast_output_parameters = ['qseqid', 'sseqid', 'evalue', 'bitscore', 'length', \n",
    "                           'pident', 'nident', 'sgi', 'sacc', 'staxids', 'sscinames', \n",
    "                           'scomnames', 'sskingdoms', 'barcode']\n",
    "for bcs, besthit_file in zip(ncbiblast_besthit_barcodes, ncbiblast_besthit_file_names):\n",
    "    if ncbiblast_file_id in besthit_file:\n",
    "        ncbiblast_dfs_dict[bcs] = pd.read_table(besthit_file, \n",
    "                                                delim_whitespace=False,\n",
    "                                                names=blast_output_parameters) #hardcoded\n",
    "for bcs in ncbiblast_dfs_dict.keys():\n",
    "    ncbiblast_dfs_dict[bcs]['barcode'] = bcs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1641,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Combining the bad barcode reads from BLAST into NB00\n",
    "if 'BC00' in ncbiblast_dfs_dict.keys():\n",
    "    ncbiblast_dfs_dict['BC00']['barcode'] = 'NB00'\n",
    "\n",
    "#Joining the dataframes for individual barcodes together\n",
    "ncbiblast_df_list = []\n",
    "for bcs in ncbiblast_dfs_dict.keys():\n",
    "    ncbiblast_df_list.append(ncbiblast_dfs_dict[bcs])\n",
    "    \n",
    "ncbiblast_df_joint = pd.concat(ncbiblast_df_list, axis=0, ignore_index=True)\n",
    "ncbiblast_df_joint.sort_values('barcode', ascending=True, inplace=True)\n",
    "ncbiblast_df_joint.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1178,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Function to generate taxonomy columns based on NCBITaxa results for NCBI hit dataframe, \n",
    "#Appends these taxcnomy columns to the joint ncbiblast dataframe\n",
    "def search_rank_output_name_append_column(df, staxid_column, rank_search):\n",
    "    \"\"\"Input df, staxid_column from same df and rank_search (a desired taxonomic rank \n",
    "    from each staxid's lineage), outputs taxonomic name corresponding to rank_search or 'Unclassified' if\n",
    "    unavailable and appends to df row by row\"\"\"\n",
    "    rank_list = []\n",
    "    for read_ind in range(0, len(staxid_column)):\n",
    "        taxid = ''\n",
    "        if ';' in str(staxid_column[read_ind]):\n",
    "            taxid = staxid_column[read_ind].split(';')[0]\n",
    "        else:\n",
    "            taxid = staxid_column[read_ind]\n",
    "        \n",
    "        taxid_lineage = ''\n",
    "        taxid_lineage = ncbi.get_lineage(taxid)\n",
    "        \n",
    "        names = ''\n",
    "        names = ncbi.get_taxid_translator(taxid_lineage)\n",
    "        \n",
    "        ranks = ''\n",
    "        ranks = ncbi.get_rank(taxid_lineage) #Dict\n",
    "        \n",
    "        ranks2names = ''\n",
    "        ranks2names = {ranks[k]:names[k] for k in names.keys() & ranks}\n",
    "        \n",
    "        if rank_search in ranks2names.keys():\n",
    "            rank_list.append(ranks2names[rank_search])#if rank in dict, print name\n",
    "        else:\n",
    "            rank_list.append('Unclassified')\n",
    "    df[rank_search] = rank_list\n",
    "# NOTE: Appending is always slow, try and find a better way e.g df.apply to a column based on staxids column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1179,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Appends columns to joint_NCBI_df_taxonomy corresponding to desired taxonomic ranks for metagenomic analysis\n",
    "ncbiblast_df_joint_taxonomy = ncbiblast_df_joint.copy()\n",
    "main_taxonomic_ranks = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "for ranks in main_taxonomic_ranks:\n",
    "    search_rank_output_name_append_column(ncbiblast_df_joint_taxonomy, \n",
    "                                          ncbiblast_df_joint_taxonomy['staxids'], \n",
    "                                          ranks)\n",
    "    \n",
    "#Removing unncessary columns from joint_NCBI_df_taxonomy\n",
    "for columns in ['scomnames', 'sskingdoms', 'sscinames']:\n",
    "    del ncbiblast_df_joint_taxonomy[columns] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1654,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Defines a function to make manual adjustments to the NCBI taxonomic sorting\n",
    "def manually_adjust_ncbi_taxonomy(rank_select, entry_select, ranks_to_edit, new_entry):\n",
    "    \"\"\"Select dataframe entries to edit by a specific rank (rank_select) \n",
    "    and entry related to that rank (entry_select), \n",
    "    provide a list of ranks to edit for those selected entries (ranks_to_edit),\n",
    "    replace these ranks with a new entry (new_entry)\"\"\"\n",
    "    if entry_select in list(ncbiblast_df_joint_taxonomy_adj[rank_select].unique()):\n",
    "        selection = (ncbiblast_df_joint_taxonomy_adj[rank_select] == entry_select)\n",
    "        selected_tax_ranks = ranks_to_edit\n",
    "        for ranks in selected_tax_ranks:\n",
    "            ncbiblast_df_joint_taxonomy_adj[ranks][selection] = new_entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1182,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "#Make two manual adjustments for this sample set\n",
    "manually_adjust_ncbi_taxonomy('species', \n",
    "                            'Cloning vector lambda TXF97', \n",
    "                            ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus'], \n",
    "                            '(Cloning vector lambda TXF97)')\n",
    "\n",
    "manually_adjust_ncbi_taxonomy('superkingdom', \n",
    "                            'Viruses', \n",
    "                            ['phylum', 'class'], \n",
    "                            'dsDNA viruses, no RNA stage')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1283,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#count pivot table of dataframe with taxonomic columns: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_ncbi_taxonomy_pivot(tax_df, rank, bcs, num):\n",
    "    \"\"\"From tax_df, generate a pivot table listing num rank counts, sorted by bcs\"\"\"\n",
    "    pivot_table = tax_df.pivot_table(values='nident', \n",
    "                                            index=rank, \n",
    "                                            columns='barcode', \n",
    "                                            aggfunc='count', \n",
    "                                            fill_value=0, \n",
    "                                            margins=True)\n",
    "    pivot_table.columns.name = None\n",
    "    return pivot_table.sort_values(bcs, axis=0, ascending=False).head(n=num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1282,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# % pivot table of dataframe with taxonomic columns: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_ncbi_taxonomy_pivot_pc(tax_df, rank, bcs, num):\n",
    "    \"\"\"From tax_df, generate a pivot table listing num rank %, sorted ascending by bcs\"\"\"\n",
    "    pivot_table_pc = tax_df.pivot_table(values='nident', \n",
    "                                        index=rank, \n",
    "                                        columns='barcode', \n",
    "                                        aggfunc='count', \n",
    "                                        fill_value=0, \n",
    "                                        margins=True)\n",
    "    pivot_table_pc.columns.name = None\n",
    "    \n",
    "    for brcs in pivot_table_pc.columns:\n",
    "        pivot_table_pc[str(brcs) + '%'] = pivot_table_pc[brcs].apply(lambda x: (x / pivot_table_pc[brcs].loc['All'])*100)\n",
    "    for brcs in pivot_table_pc.columns:\n",
    "        if '%' not in brcs:\n",
    "            pivot_table_pc.drop(brcs, axis=1, inplace=True)\n",
    "    pivot_table_pc.drop('All', axis=0, inplace=True)\n",
    "\n",
    "    pivot_table_pc2 = pivot_table_pc.sort_values(str(bcs) + '%', axis=0, ascending=False)\n",
    "    return pivot_table_pc2.applymap(lambda x: '%.3f' % x).head(n=num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1300,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#count pivot table of dataframe with taxonomic columns, includes reference genomes set across all ranks: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_all_hits_pivot(rank, bcs, num, dropmaxspecies=False):\n",
    "    \"\"\"generate a pivot table listing num rank counts, sorted ascending by bcs\"\"\"\n",
    "    blast_and_NCBI_df_list = [generate_ncbi_taxonomy_pivot(ncbiblast_df_joint_taxonomy_adj, rank, bcs, len(ncbiblast_df_joint_taxonomy_adj)), \n",
    "                    rgblast_df_joint_pivot.transpose()]\n",
    "    blast_and_NCBI_hits = pd.concat(blast_and_NCBI_df_list)\n",
    "    blast_and_NCBI_hits_sum = blast_and_NCBI_hits.groupby(blast_and_NCBI_hits.index).sum()\n",
    "    \n",
    "    max_genome = str(rgblast_df_joint_pivot_noall.ix['All'].idxmax(axis=1))\n",
    "    \n",
    "    if dropmaxspecies == True:\n",
    "        blast_and_NCBI_hits_sum.loc['All'] = blast_and_NCBI_hits_sum.loc['All'] - blast_and_NCBI_hits_sum.loc[max_genome]\n",
    "        blast_and_NCBI_hits_sum.drop(max_genome, axis=0, inplace=True)\n",
    "    \n",
    "    return blast_and_NCBI_hits_sum.sort_values(bcs, axis=0, ascending=False).head(n=num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1301,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# % pivot table of dataframe with taxonomic columns, includes reference genomes set across all ranks: \n",
    "#rows - rank names, columns - barcode, sorted by desired class\n",
    "def generate_all_hits_pivot_pc(rank, bcs, num, dropmaxspecies=False):\n",
    "    \"\"\"generate a pivot table listing num rank %, sorted ascending by bcs\"\"\"\n",
    "    blast_and_NCBI_df_list = [generate_ncbi_taxonomy_pivot(ncbiblast_df_joint_taxonomy_adj, rank, bcs, len(ncbiblast_df_joint_taxonomy_adj)), \n",
    "                    rgblast_df_joint_pivot.transpose()]\n",
    "    blast_and_NCBI_hits = pd.concat(blast_and_NCBI_df_list)\n",
    "    blast_and_NCBI_hits_sum = blast_and_NCBI_hits.groupby(blast_and_NCBI_hits.index).sum()\n",
    "    \n",
    "    max_genome = str(rgblast_df_joint_pivot_noall.ix['All'].idxmax(axis=1))\n",
    "    \n",
    "    if dropmaxspecies == True:\n",
    "        blast_and_NCBI_hits_sum.loc['All'] = blast_and_NCBI_hits_sum.loc['All'] - blast_and_NCBI_hits_sum.loc[max_genome]\n",
    "        blast_and_NCBI_hits_sum.drop(max_genome, axis=0, inplace=True)\n",
    "    \n",
    "    for brcs in blast_and_NCBI_hits_sum.columns:\n",
    "        blast_and_NCBI_hits_sum[str(brcs) + '%'] = blast_and_NCBI_hits_sum[brcs].apply(lambda x: (x / blast_and_NCBI_hits_sum[brcs].loc['All'])*100)\n",
    "    for brcs in blast_and_NCBI_hits_sum.columns:\n",
    "        if '%' not in brcs:\n",
    "            blast_and_NCBI_hits_sum.drop(brcs, axis=1, inplace=True)\n",
    "    blast_and_NCBI_hits_sum.drop('All', axis=0, inplace=True)\n",
    "    \n",
    "    blast_and_NCBI_hits_sum2 = blast_and_NCBI_hits_sum.sort_values(str(bcs) + '%', axis=0, ascending=False)\n",
    "    return blast_and_NCBI_hits_sum2.applymap(lambda x: '%.3f' % x).head(n=num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1526,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "#Assesses number of barcodes, generates necessary number of plots (even) and relevant indexing system\n",
    "ncbiblast_barcodes = list(ncbiblast_df_joint_taxonomy_adj['barcode'].unique())\n",
    "\n",
    "if len(list(ncbiblast_df_joint_taxonomy_adj['barcode'].unique())) % 2 == 0:\n",
    "    no_of_subplots = len(ncbiblast_barcodes)\n",
    "else:\n",
    "    no_of_subplots = len(ncbiblast_barcodes) + 1\n",
    "\n",
    "#Always 2 columns, bc-count/2 rows \n",
    "no_of_subplots_pair = [int(no_of_subplots/2), 2]\n",
    "\n",
    "#Produce pairs of indices correlating to the coordinates of the subplots\n",
    "subplot_coordinates = list(product(range(no_of_subplots_pair[0]), range(no_of_subplots_pair[1])))\n",
    "subplot_coordinates_list = [list(l) for l in subplot_coordinates]\n",
    "subplot_coordinates_list_rows = [i[0] for i in subplot_coordinates_list]\n",
    "subplot_coordinates_list_columns = [i[1] for i in subplot_coordinates_list]\n",
    "\n",
    "fig, ax = plt.subplots(no_of_subplots_pair[1], no_of_subplots_pair[0], figsize=(20, 6))\n",
    "\n",
    "def applyTaxPlotStyle(ax_ind1, ax_ind2, bcs, rank):\n",
    "    labels = ''\n",
    "    labels = tuple(generate_ncbi_taxonomy_pivot_pc(ncbiblast_df_joint_taxonomy_adj, rank, bcs, 10)[bcs + '%'].index)\n",
    "    y_pos = ''\n",
    "    y_pos = np.arange(len(labels))\n",
    "    sizes = ''\n",
    "    sizes = tuple([float(i) for i in generate_ncbi_taxonomy_pivot_pc(ncbiblast_df_joint_taxonomy_adj, rank, bcs, 10)[bcs + '%']])\n",
    "\n",
    "    ax[ax_ind1, ax_ind2].barh(y_pos, sizes, align='center', alpha=0.4, color=random_color(x))\n",
    "    ax[ax_ind1, ax_ind2].set_yticks(y_pos)\n",
    "    ax[ax_ind1, ax_ind2].set_yticklabels(labels)\n",
    "    ax[ax_ind1, ax_ind2].tick_params(top=\"off\", right=\"off\", left=\"off\")\n",
    "    ax[ax_ind1, ax_ind2].set_xlabel('% Total Reads Hitting NCBI')\n",
    "    ax[ax_ind1, ax_ind2].set_xticks(range(0, 110, 10))\n",
    "    ax[ax_ind1, ax_ind2].set_title(bcs)\n",
    "    ax[ax_ind1, ax_ind2].invert_yaxis()\n",
    "    ax[ax_ind1, ax_ind2].grid(b=None, which='major', axis='x', linestyle='dashed', linewidth=0.2)\n",
    "    ax[ax_ind1, ax_ind2].annotate('E = 0.00001', \n",
    "             xy=(1.06, 0.3), xycoords='axes fraction', fontsize=10, xytext=(-20, -30), textcoords='offset points', \n",
    "             ha='right', va='bottom')\n",
    "    \n",
    "def generateTaxSubplots(search_rank):\n",
    "    for ax_ind1, ax_ind2, bcs, in zip(subplot_coordinates_list_columns, subplot_coordinates_list_rows, bc_list):\n",
    "        applyTaxPlotStyle(ax_ind1, ax_ind2, bcs, search_rank)    \n",
    "\n",
    "    if len(list(ncbiblast_df_joint_taxonomy_adj['barcode'].unique())) != 0:\n",
    "        plt.delaxes(ax[subplot_coordinates_list_columns[-1], subplot_coordinates_list_rows[-1]])\n",
    "    \n",
    "    plt.suptitle('Reads Hitting NCBI Database - % By ' + str(search_rank).title() + ' By Barcode', \n",
    "             fontsize='x-large',\n",
    "             y=1.02,\n",
    "             fontweight=\"bold\")\n",
    "    \n",
    "    plt.subplots_adjust(left=None, bottom=None, right=None, top=None, wspace=1.4, hspace=0.4)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ncbi_hit_percentage_' + search_rank + '.png', bbox_inches='tight')\n",
    "            \n",
    "generateTaxSubplots('genus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1527,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "fig, ax = plt.subplots(no_of_subplots_pair[1], no_of_subplots_pair[0], figsize=(20, 6))\n",
    "generateTaxSubplots('species')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1528,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "#Assesses number of barcodes, generates necessary number of plots (even) and relevant indexing system\n",
    "ncbiblast_barcodes = list(ncbiblast_df_joint_taxonomy_adj['barcode'].unique())\n",
    "\n",
    "if len(list(ncbiblast_df_joint_taxonomy_adj['barcode'].unique())) % 2 == 0:\n",
    "    no_of_subplots = len(ncbiblast_barcodes)\n",
    "else:\n",
    "    no_of_subplots = len(ncbiblast_barcodes) + 1\n",
    "\n",
    "#Always 2 columns, bc-count/2 rows \n",
    "no_of_subplots_pair = [int(no_of_subplots/2), 2]\n",
    "\n",
    "#Produce pairs of indices correlating to the coordinates of the subplots\n",
    "subplot_coordinates = list(product(range(no_of_subplots_pair[0]), range(no_of_subplots_pair[1])))\n",
    "subplot_coordinates_list = [list(l) for l in subplot_coordinates]\n",
    "subplot_coordinates_list_rows = [i[0] for i in subplot_coordinates_list]\n",
    "subplot_coordinates_list_columns = [i[1] for i in subplot_coordinates_list]\n",
    "\n",
    "fig, ax = plt.subplots(no_of_subplots_pair[1], no_of_subplots_pair[0], figsize=(20, 6))\n",
    "max_genome = str(rgblast_df_joint_pivot_noall.ix['All'].idxmax(axis=1))\n",
    "\n",
    "def applyTaxPlotStyle(ax_ind1, ax_ind2, bcs, rank):\n",
    "    cycol_one = ''\n",
    "    cycol_one = random_color(x)\n",
    "    labels = ''\n",
    "    labels = tuple(generate_all_hits_pivot_pc(rank, bcs, 10, dropmaxspecies=False)[bcs + '%'].index)\n",
    "    y_pos = ''\n",
    "    y_pos = np.arange(len(labels))\n",
    "    sizes = ''\n",
    "    sizes = tuple([float(i) for i in generate_all_hits_pivot_pc(rank, bcs, 10, dropmaxspecies=False)[bcs + '%']])\n",
    "\n",
    "    colors = []\n",
    "    for name in labels:\n",
    "        if name in list(rgblast_df_joint_pivot.columns):\n",
    "            colors.append('k')\n",
    "        else:\n",
    "            colors.append(cycol_one)\n",
    "    \n",
    "    ax[ax_ind1, ax_ind2].barh(y_pos, sizes, align='center', alpha=0.4, color=colors)\n",
    "    ax[ax_ind1, ax_ind2].set_yticks(y_pos)\n",
    "    ax[ax_ind1, ax_ind2].set_yticklabels(labels)\n",
    "    ax[ax_ind1, ax_ind2].tick_params(top=\"off\", right=\"off\", left=\"off\")\n",
    "    \n",
    "    exact_xlim = int(float(generate_all_hits_pivot_pc(rank, 'BC04', len(ncbiblast_df_joint_taxonomy_adj), dropmaxspecies=False).drop(max_genome).values.max())+1)\n",
    "    \n",
    "    ax[ax_ind1, ax_ind2].set_xlim([0, exact_xlim])\n",
    "    ax[ax_ind1, ax_ind2].set_xlabel('% Total Reads Hitting Blast or NCBI')\n",
    "    ax[ax_ind1, ax_ind2].set_xticks(range(0, exact_xlim+1, 1))\n",
    "    ax[ax_ind1, ax_ind2].set_title(bcs)\n",
    "    ax[ax_ind1, ax_ind2].invert_yaxis()\n",
    "    ax[ax_ind1, ax_ind2].grid(b=None, which='major', axis='x', linestyle='dashed', linewidth=0.2)\n",
    "    ax[ax_ind1, ax_ind2].annotate(max_genome + ' = ' + str(sizes[0]) + '%', \n",
    "             xy=(1.06, 0.3), xycoords='axes fraction', fontsize=10, xytext=(-20, -30), textcoords='offset points', \n",
    "             ha='right', va='bottom')\n",
    "    ax[ax_ind1, ax_ind2].annotate('E = 0.00001', \n",
    "             xy=(1.06, 0.4), xycoords='axes fraction', fontsize=10, xytext=(-20, -30), textcoords='offset points', \n",
    "             ha='right', va='bottom')\n",
    "    \n",
    "def generateTaxRGSubplots(search_rank):\n",
    "    for ax_ind1, ax_ind2, bcs, in zip(subplot_coordinates_list_columns, subplot_coordinates_list_rows, bc_list):\n",
    "        applyTaxPlotStyle(ax_ind1, ax_ind2, bcs, search_rank)    \n",
    "\n",
    "    if len(list(ncbiblast_df_joint_taxonomy_adj['barcode'].unique())) != 0:\n",
    "        plt.delaxes(ax[subplot_coordinates_list_columns[-1], subplot_coordinates_list_rows[-1]])\n",
    "    \n",
    "    plt.suptitle('Reads Hitting Reference Genomes or NCBI Database - % By ' + str(search_rank).title() + ' By Barcode', \n",
    "             fontsize='x-large',\n",
    "             y=1.02,\n",
    "             fontweight=\"bold\")\n",
    "    \n",
    "    plt.subplots_adjust(left=None, bottom=None, right=None, top=None, wspace=1.4, hspace=0.4)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ncbi_and_ref_genome_hit_percentage_' + search_rank + '.png', bbox_inches='tight')\n",
    "\n",
    "generateTaxRGSubplots('genus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1529,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "fig, ax = plt.subplots(no_of_subplots_pair[1], no_of_subplots_pair[0], figsize=(20, 7))\n",
    "generateTaxRGSubplots('species')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1443,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1212_1D_barcoding_Wagga_\n",
      ".blasthitsample_\n",
      ".blasthit\n",
      ".NCBI04052016\n",
      ".WW_19121016\n",
      ".fa\n",
      ".blast.besthit\n"
     ]
    }
   ],
   "source": [
    "print(basecalled_fasta_file_id)\n",
    "print(blast_hit_sample_id)\n",
    "norgblasted_hit_id = '.noblasthit'\n",
    "print(rgblasted_hit_id)\n",
    "print(ncbiblast_file_id)\n",
    "print(rgblast_file_id)\n",
    "print(fasta_file_type)\n",
    "print(rgblast_file_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1458,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def generate_ncbihit_fasta(bcs, ncbihit):\n",
    "    \"\"\"Takes basecalled.df, and list of read_ids from ncbiblast_df_joint. For a single barcode:\n",
    "    - if rghit is yes, extract a fasta file with all the reads hitting the reference genomes to data/rghityes\n",
    "    - if rghit is no, extract a fasta file with all the reads not hitting the reference genomes to data/rghitno\"\"\"\n",
    "    os.chdir(work_dir + 'data/rghitno/')\n",
    "    read_id_list = basecalled_df.groupby('Barcode')['Read_id'].get_group(bcs).tolist()\n",
    "    ncbi_qseqid_list = ncbiblast_df_joint.groupby('barcode')['qseqid'].get_group(bcs).tolist()\n",
    "    norgblasthit_file = basecalled_fasta_file_id + bcs + rgblast_file_id + norgblasted_hit_id + fasta_file_type\n",
    "    \n",
    "    if ncbihit == 'Yes':\n",
    "        ncbiblasted_hit_id = '.blasthit'\n",
    "        ncbiblasthit_id = ncbiblast_file_id + ncbiblasted_hit_id + fasta_file_type\n",
    "        ncbiblasthit_file = basecalled_fasta_file_id + bcs + rgblast_file_id + rgblast_file_type + ncbiblasthit_id\n",
    "\n",
    "        ncbihityes = ''\n",
    "        ncbihityes = (rec for rec in SeqIO.parse(norgblasthit_file, 'fasta') if rec.id in ncbi_qseqid_list) \n",
    "        count = ''\n",
    "        count = SeqIO.write(ncbihityes, os.path.join(work_dir + 'data/rghitno/ncbihityes', ncbiblasthit_file), \"fasta\")\n",
    "            \n",
    "    if ncbihit == 'No':\n",
    "        noncbiblasted_hit_id = '.noblasthit'\n",
    "        noncbiblasthit_id = ncbiblast_file_id + noncbiblasted_hit_id + fasta_file_type\n",
    "        noncbiblasthit_file = basecalled_fasta_file_id + bcs + rgblast_file_id + rgblast_file_type + noncbiblasthit_id\n",
    "        \n",
    "        ncbihitno = ''\n",
    "        ncbihitno = (rec for rec in SeqIO.parse(norgblasthit_file, 'fasta') if rec.id not in ncbi_qseqid_list) \n",
    "        count = ''\n",
    "        count = SeqIO.write(ncbihitno, os.path.join(work_dir + 'data/rghitno/ncbihitno', noncbiblasthit_file), \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1452,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "450\n",
      "4127\n",
      "1764\n",
      "8746\n",
      "4424\n",
      "3212\n",
      "181\n"
     ]
    }
   ],
   "source": [
    "for bcs in list(ncbiblast_df_joint.barcode.unique()):\n",
    "    generate_ncbihit_fasta(bcs, 'Yes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1453,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10029\n",
      "8386\n",
      "1543\n",
      "3541\n",
      "3435\n",
      "26083\n",
      "5543\n"
     ]
    }
   ],
   "source": [
    "for bcs in list(ncbiblast_df_joint.barcode.unique()):\n",
    "    generate_ncbihit_fasta(bcs, 'No')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1473,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ncbiblast_df_joint_pivot = ncbiblast_df_joint.pivot_table(values='nident', index='barcode', columns='staxids', aggfunc='count', fill_value=0, margins=True)\n",
    "ncbiblast_df_joint_pivot = ncbiblast_df_joint_pivot.applymap(int)\n",
    "ncbiblast_df_joint_pivot.index.name = None\n",
    "ncbiblast_df_joint_pivot.columns.name = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1510,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ncbiblasted_count_df = pd.concat([rgblasted_count_df['RGHitNo'], ncbiblast_df_joint_pivot['All']],  axis=1)\n",
    "ncbiblasted_count_df.applymap(lambda x: int(x))\n",
    "ncbiblasted_count_df.rename(columns={'RGHitNo' : 'NCBICount', 'All' : 'NCBIHitYes'}, inplace=True)\n",
    "ncbiblasted_count_df['NCBIHitNo'] = ncbiblasted_count_df['NCBICount'] - ncbiblasted_count_df['NCBIHitYes']\n",
    "ncbiblasted_count_df['NCBIHitYes%'] = np.round((ncbiblasted_count_df['NCBIHitYes'] / ncbiblasted_count_df['NCBICount'] * 100), 3)\n",
    "ncbiblasted_count_df['NCBIHitNo%'] = np.round((ncbiblasted_count_df['NCBIHitNo'] / ncbiblasted_count_df['NCBICount'] * 100), 3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1511,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ncbiblasted_count_df_noall = ncbiblasted_count_df.drop('All')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1512,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Saves pivot table summary in analysis/tables/ folder ###\n",
    "fig = plt.figure(figsize=(10,2))\n",
    "ax = fig.add_subplot(1, 1, 1, frame_on=False)\n",
    "ax.xaxis.set_visible(False)  # hide the x axis\n",
    "ax.yaxis.set_visible(False)  # hide the y axis\n",
    "\n",
    "table(ax, ncbiblasted_count_df, loc='center')\n",
    "\n",
    "plt.savefig(os.path.join(work_dir + 'analysis/tables/', \n",
    "                        basecalled_fasta_file_id + \"ncbiblast_hit_vs_nohit_summary.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1530,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "#two qualitative variables - barcode, blasted vs no hit vs hit\n",
    "#one quantitative - read count\n",
    "n_groups = len(ncbiblasted_count_df_noall.index.values)\n",
    "\n",
    "bar_labels = ['Blasted', 'Blasted and Hit', 'Blasted and No Hit']\n",
    "counts = [tuple(ncbiblasted_count_df_noall['NCBICount']),\n",
    "          tuple(ncbiblasted_count_df_noall['NCBIHitYes']),\n",
    "          tuple(ncbiblasted_count_df_noall['NCBIHitNo'])]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6.5,3))\n",
    "\n",
    "index = np.arange(n_groups)\n",
    "bar_width = 0.15\n",
    "opacity = 0.4\n",
    "\n",
    "for idx, (lbl, cnts) in enumerate(zip(bar_labels, counts)):\n",
    "    plt.bar(index + bar_width*idx, cnts, bar_width,\n",
    "                 alpha=opacity,\n",
    "                 color=random_color(x),\n",
    "                 label=lbl)\n",
    "    \n",
    "plt.xlabel('Barcode')\n",
    "plt.ylabel('Read Count')\n",
    "plt.title('Number of Reads Blasted and Hit To Reference Genomes', y=1.08)\n",
    "plt.xticks(index + bar_width*1.5, tuple(ncbiblasted_count_df_noall.index.values))\n",
    "plt.tick_params(right='off', top='off', bottom='off')\n",
    "plt.legend(loc='center left', bbox_to_anchor=(1.01, 0.5), fontsize=10)\n",
    "plt.grid(b=None, which='major', axis='y', linestyle='dashed', linewidth=0.2)\n",
    "plt.margins(x=0.1)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ncbi_hit_and_nohit_counts.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1531,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamran/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "#two qualitative variables - barcode, no hit vs hit\n",
    "#one quantitative - read count\n",
    "n_groups = len(ncbiblasted_count_df_noall.index.values)\n",
    "\n",
    "bar_labels = ['Blasted and Hit', 'Blasted and No Hit']\n",
    "counts = [tuple(ncbiblasted_count_df_noall['NCBIHitYes%']),\n",
    "          tuple(ncbiblasted_count_df_noall['NCBIHitNo%'])]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6.5,3))\n",
    "\n",
    "index = np.arange(n_groups)\n",
    "bar_width = 0.15\n",
    "opacity = 0.4\n",
    "\n",
    "for idx, (lbl, cnts) in enumerate(zip(bar_labels, counts)):\n",
    "    plt.bar(index + bar_width*idx, cnts, bar_width,\n",
    "                 alpha=opacity,\n",
    "                 color=random_color(x),\n",
    "                 label=lbl)\n",
    "\n",
    "plt.xlabel('Barcode')\n",
    "plt.ylabel('% of Reads Per Barcode')\n",
    "plt.title('% of Reads Blasted and Hit To Reference Genomes', y=1.08)\n",
    "plt.xticks(index + bar_width, tuple(ncbiblasted_count_df_noall.index.values), y=-0.01)\n",
    "plt.tick_params(top='off', right='off')\n",
    "plt.legend(loc='center left', bbox_to_anchor=(1.01, 0.5), fontsize=12)\n",
    "plt.grid(b=None, which='major', axis='y', linestyle='dashed', linewidth=0.2)\n",
    "plt.margins(x=0.1)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(work_dir + 'analysis/graphs/' + basecalled_fasta_file_id + 'ncbi_hit_and_nohit_percentage.png', bbox_inches='tight')"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
